---
title: "RNA-seq analysis in R"
subtitle: "Differential Expression of RNA-seq data - Solutions"
author: "Stephane Ballereau, Dominique-Laurent Couturier, Mark Dunning, Abbi Edwards, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float: yes
minutes: 300
layout: page
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(tidyverse)
library(DESeq2)
```

```{r prepareData, include=F}
load("../Robjects/preprocessing.RData")
# Specify a model
design <- as.formula(~ CellType + Status)
# set virgin to base line
sampleinfo$Status <- factor(sampleinfo$Status, 
                              levels = c("virgin", "pregnant", "lactate"))
# rebuild a clean DDS object
ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = design)
# Run DESeq
ddsObj <- DESeq(ddsObj)
```


### Challenge 1

Obtain results for luminal vs basal and find the top 200 genes (call the new
results object `resLvB`.

```{r solutionChallenge1}
resLvB <- results(ddsObj, name="CellType_luminal_vs_basal", alpha=0.05)
topGenesLvB <- as.data.frame(resLvB) %>%
    rownames_to_column("GeneID") %>% 
    arrange(padj) %>%
    head(200)
head(topGenesLvB, 10)
```

### Challenge 2

When we looked at the PCA it did seem that an interaction model might be
warranted. Let's test that.   

> 1. Fit a model with interaction: What is the rationale to include the 
> interaction?  
> 2. Use the LRT to compare the two models.  
> 3. Is the number of replicates good enough to include the interaction?  
> 4. Is the interaction needed in the model?  

```{r solutionChallenge2}
designI <- as.formula(~ CellType * Status)

# Build model
ddsObjI <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = designI)
# Run DESeq
ddsObjI <- DESeq(ddsObjI)

# Compare the models
ddsObjI <- DESeq(ddsObjI, test="LRT", reduced=design)
resIvA <- results(ddsObjI)
head(resIvA)
length(which(resIvA$padj<=0.05))
```
