---
title: "RNA-seq Analysis in R"
subtitle: "Annotation and Visualisation of RNA-seq results - Solutions"
author: "Stephane Ballereau, Mark Dunning, Oscar Rueda, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
    toc_float: yes
  html_document:
    toc: yes
    toc_float: yes
minutes: 300
layout: page
---

```{r setup, include=FALSE}
library(biomaRt)
library(GenomicFeatures)
library(DESeq2)
library(tidyverse)
```

```{r prepareData, cache=TRUE}
# load data
load("../Robjects/DE.RData")
## set up connection to ensembl database
ensembl=useMart("ENSEMBL_MART_ENSEMBL")
# list the available datasets (species)
listDatasets(ensembl) %>% 
    filter(str_detect(description, "Mouse"))
# specify a data set to use
ensembl = useDataset("mmusculus_gene_ensembl", mart=ensembl)
# Set the filter type and values
filterType <- "ensembl_gene_id"
# annotated
load("../Robjects/Ensembl_annotations.RData")
annotLvV <- as.data.frame(resLvV) %>% 
    rownames_to_column("GeneID") %>% 
    left_join(ensemblAnnot, "GeneID") %>% 
    rename(logFC=log2FoldChange, FDR=padj)
# txdb object
txMm <- makeTxDbFromBiomart(dataset="mmusculus_gene_ensembl")
```


> ### Challenge 1 {.challenge}
> That was just 1000 genes. We need annotations for the entire results table.
> Also, there may be some other interesting columns in BioMart that we wish to
> retrieve.  
>
> (a) Search the attributes and add the following to our list of attributes:  
>       (i) The gene description   
>       (ii) The genomic position - chromosome, start, end, and strand (4 columns) 
>       (iii) The gene biotype  
> (b) Query BioMart using all of the genes in our results table (`resLvV`)  
> (c) How many Ensembl genes have multipe Entrez IDs associated with them?  
> (d) How many Ensembl genes in `resLvV` don't have any annotation? Why is this?

```{r solutionChallenge1}
filterValues <- rownames(resLvV)

# check the available "attributes" - things you can retreive
listAttributes(ensembl) %>%
    head(20)
attributeNames <- c('ensembl_gene_id',
                    'entrezgene',
                    'external_gene_name',
                    'description',
                    'gene_biotype')

# run the query
annot <- getBM(attributes=attributeNames,
               filters = ourFilterType,
               values = filterValues,
               mart = ensembl)

# dulicate ids
sum(duplicated(annot$ensembl_gene_id))

# missing gens
missingGenes <- !rownames(resLvV)%in%annot$ensembl_gene_id
rownames(resLvV)[missingGenes]
```


> ### Challenge 2 {.challenge}

> Use the log2 fold change (`logFC`) on the x-axis, and use `-log10(FDR)` on the y-axis. (This >`-log10` transformation is commonly used for p-values as it means that more significant genes have a >higher scale) 
>
> (a) Create a column of -log10(FDR) values
>
> (b) Create a plot with points coloured by if FDR < 0.05


```{r volcanoPlot, fig.height=5, fig.width=7}
# first remove the filtered genes (FDR=NA) and create a -log10(FDR) column
filtTab <- shrinkLvV %>% 
    filter(!is.na(FDR)) %>% 
    mutate(`-log10(FDR)` = -log10(FDR))

ggplot(filtTab, aes(x = logFC, y=`-log10(FDR)`)) + 
    geom_point(aes(colour=FDR < 0.05), size=1)
```


> ### Challenge 3 {.challenge} - In Supplementary Materials
>
> Use the txMm to retrieve the exon coordinates for the genes: 
    + `ENSMUSG00000021604`
    + `ENSMUSG00000022146`
    + `ENSMUSG00000040118` 
>

```{r solutionChallenge3, echo=FALSE, warning=FALSE, message=FALSE}
keyList <- c("ENSMUSG00000021604", "ENSMUSG00000022146", "ENSMUSG00000040118")
select(txMm, 
       keys=keyList,
       keytype = "GENEID",
       columns=c("TXNAME", "TXCHROM", "TXSTART", "TXEND", "TXSTRAND", "TXTYPE")
      )
```
