<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>RNA-seq analysis in R</title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="Software Carpentry Blog" href="http://software-carpentry.org/feed.xml"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <a href="index.html"><h1 class="title">RNA-seq analysis in R</h1></a>
          <h2 class="subtitle">Differential expression analysis</h2>
          <p><strong>Authors: Belinda Phipson, Anna Trigos, Matt Ritchie, Maria Doyle, Harriet Dashnow, Charity Law</strong></p>
<h2 id="resources-and-data-files">Resources and data files</h2>
<p>This material has been created using the following resources:<br />
http://www.statsci.org/smyth/pubs/QLedgeRPreprint.pdf <span class="citation">(Lun, Chen, and Smyth 2016)</span><br />
http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/99-RNAseq_DE_analysis_with_R.html</p>
<p>Data files downloaded from:<br />
ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE60nnn/GSE60450/suppl/GSE60450_Lactation-GenewiseCounts.txt.gz http://bioinf.wehi.edu.au/software/MSigDB/mouse_c2_v5.rdata http://bioinf.wehi.edu.au/software/MSigDB/mouse_H_v5.rdata</p>
<p>Data files:<br />
sampleinfo.txt<br />
GSE60450_Lactation-GenewiseCounts.txt<br />
mouse_c2_v5.rdata<br />
mouse_H_v5.rdata</p>
<p>Data files available from: <a href="https://figshare.com/s/1d788fd384d33e913a2a" class="uri">https://figshare.com/s/1d788fd384d33e913a2a</a> You should download these files and place them in your <code>/data</code> directory.</p>
<p>Packages used: limma, edgeR, gplots, org.Mm.eg.db, RColorBrewer, Glimma</p>
<h2 id="overview">Overview</h2>
<ul>
<li>Reading in table of counts</li>
<li>Filtering lowly expressed genes</li>
<li>Quality control</li>
<li>Normalisation for composition bias</li>
<li>Differential expression analysis</li>
<li>Testing relative to a threshold</li>
<li>Visualisation</li>
<li>Gene set testing</li>
</ul>
<h2 id="introduction-and-data-import">Introduction and data import</h2>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA-Seq to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA-Seq experiment. Analysing an RNAseq experiment begins with sequencing reads. These are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and getting stuck into analysis.</p>
<p>First, let’s load all the packages we will need to analyse the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(edgeR)
<span class="kw">library</span>(limma)
<span class="kw">library</span>(Glimma)
<span class="kw">library</span>(gplots)
<span class="kw">library</span>(org.Mm.eg.db)
<span class="kw">library</span>(RColorBrewer)</code></pre></div>
<h3 id="mouse-mammary-gland-dataset">Mouse mammary gland dataset</h3>
<p>The data for this tutorial comes from a Nature Cell Biology paper, <a href="http://www.ncbi.nlm.nih.gov/pubmed/25730472"><em>EGF-mediated induction of Mcl-1 at the switch to lactation is essential for alveolar cell survival</em></a> <span class="citation">(Fu et al. 2015)</span>. Both the raw data (sequence reads) and processed data (counts) can be downloaded from Gene Expression Omnibus database (GEO) under accession number <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450">GSE60450</a>.</p>
<p>This study examines the expression profiles of basal stem-cell enriched cells (B) and committed luminal cells (L) in the mammary gland of virgin, pregnant and lactating mice. Six groups are present, with one for each combination of cell type and mouse status. Each group contains two biological replicates. We will first use the counts file as a starting point for our analysis. This data has already been aligned to the mouse genome. The command line tool featureCounts <span class="citation">(Liao, Smyth, and Shi 2014)</span> was used to count reads mapped to mouse genes from Refseq annotation (see the <a href="http://www.ncbi.nlm.nih.gov/pubmed/25730472">paper</a> for details).</p>
<h3 id="reading-in-the-data">Reading in the data</h3>
<p><em>Set up an RStudio project specifying the directory where you have saved the <code>/data</code> directory</em>. Download and read in the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Read the data into R</span>
seqdata &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/GSE60450_Lactation-GenewiseCounts.txt&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
<span class="co"># Read the sample information into R</span>
sampleinfo &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/SampleInfo.txt&quot;</span>)</code></pre></div>
<p>Let’s take a look at the data. You can use the <code>head</code> command to see the first 6 lines. The <code>dim</code> command will tell you how many rows and columns the data frame has.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(seqdata)</code></pre></div>
<pre class="output"><code>  EntrezGeneID Length MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1
1       497097   3634                               438
2    100503874   3259                                 1
3    100038431   1634                                 0
4        19888   9747                                 1
5        20671   3130                               106
6        27395   4203                               309
  MCL1.DH_BC2CTUACXX_CAGATC_L002_R1 MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1
1                               300                                65
2                                 0                                 1
3                                 0                                 0
4                                 1                                 0
5                               182                                82
6                               234                               337
  MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1 MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1
1                               237                               354
2                                 1                                 0
3                                 0                                 0
4                                 0                                 0
5                               105                                43
6                               300                               290
  MCL1.DL_BC2CTUACXX_ATCACG_L002_R1 MCL1.LA_BC2CTUACXX_GATCAG_L001_R1
1                               287                                 0
2                                 4                                 0
3                                 0                                 0
4                                 0                                10
5                                82                                16
6                               270                               560
  MCL1.LB_BC2CTUACXX_TGACCA_L001_R1 MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1
1                                 0                                 0
2                                 0                                 0
3                                 0                                 0
4                                 3                                10
5                                25                                18
6                               464                               489
  MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1 MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1
1                                 0                                 0
2                                 0                                 0
3                                 0                                 0
4                                 2                                 0
5                                 8                                 3
6                               328                               307
  MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1
1                                 0
2                                 0
3                                 0
4                                 0
5                                10
6                               342
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(seqdata)</code></pre></div>
<pre class="output"><code>[1] 27179    14
</code></pre>
<p>The <code>seqdata</code> object contains information about genes (one gene per row), the first column has the Entrez gene id, the second has the gene length and the remaining columns contain information about the number of reads aligning to the gene in each experimental sample. There are two replicates for each cell type and timepoint (detailed sample info can be found in file “GSE60450_series_matrix.txt” from the <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450">GEO website</a>). The sampleinfo file contains basic information about the samples that we will need for the analysis today.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sampleinfo</code></pre></div>
<pre class="output"><code>                            FileName SampleName CellType   Status
1  MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1    MCL1.DG  luminal   virgin
2  MCL1.DH_BC2CTUACXX_CAGATC_L002_R1    MCL1.DH    basal   virgin
3  MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1    MCL1.DI    basal pregnant
4  MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1    MCL1.DJ    basal pregnant
5  MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1    MCL1.DK    basal  lactate
6  MCL1.DL_BC2CTUACXX_ATCACG_L002_R1    MCL1.DL    basal  lactate
7  MCL1.LA_BC2CTUACXX_GATCAG_L001_R1    MCL1.LA    basal   virgin
8  MCL1.LB_BC2CTUACXX_TGACCA_L001_R1    MCL1.LB  luminal   virgin
9  MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1    MCL1.LC  luminal pregnant
10 MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1    MCL1.LD  luminal pregnant
11 MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1    MCL1.LE  luminal  lactate
12 MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1    MCL1.LF  luminal  lactate
</code></pre>
<p>We will be manipulating and reformating the counts matrix into a suitable format for downstream analysis. The first two columns in the <code>seqdata</code> dataframe contain annotation information. We need to make a new matrix containing only the counts, but we can store the gene identifiers (the <code>EntrezGeneID</code> column) as rownames. We will add more annotation information about each gene later on in the workshop.</p>
<h3 id="format-the-data">Format the data</h3>
<p>Let’s create a new data object, <code>countdata</code>, that contains only the counts for the 12 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Remove first two columns from seqdata</span>
countdata &lt;-<span class="st"> </span>seqdata[,-(<span class="dv">1</span>:<span class="dv">2</span>)]
<span class="co"># Look at the output</span>
<span class="kw">head</span>(countdata)</code></pre></div>
<pre class="output"><code>  MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1 MCL1.DH_BC2CTUACXX_CAGATC_L002_R1
1                               438                               300
2                                 1                                 0
3                                 0                                 0
4                                 1                                 1
5                               106                               182
6                               309                               234
  MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1 MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1
1                                65                               237
2                                 1                                 1
3                                 0                                 0
4                                 0                                 0
5                                82                               105
6                               337                               300
  MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1 MCL1.DL_BC2CTUACXX_ATCACG_L002_R1
1                               354                               287
2                                 0                                 4
3                                 0                                 0
4                                 0                                 0
5                                43                                82
6                               290                               270
  MCL1.LA_BC2CTUACXX_GATCAG_L001_R1 MCL1.LB_BC2CTUACXX_TGACCA_L001_R1
1                                 0                                 0
2                                 0                                 0
3                                 0                                 0
4                                10                                 3
5                                16                                25
6                               560                               464
  MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1 MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1
1                                 0                                 0
2                                 0                                 0
3                                 0                                 0
4                                10                                 2
5                                18                                 8
6                               489                               328
  MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1 MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1
1                                 0                                 0
2                                 0                                 0
3                                 0                                 0
4                                 0                                 0
5                                 3                                10
6                               307                               342
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Store EntrezGeneID as rownames</span>
<span class="kw">rownames</span>(countdata) &lt;-<span class="st"> </span>seqdata[,<span class="dv">1</span>]</code></pre></div>
<p>Take a look at the output</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(countdata)</code></pre></div>
<pre class="output"><code>          MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1
497097                                  438
100503874                                 1
100038431                                 0
19888                                     1
20671                                   106
27395                                   309
          MCL1.DH_BC2CTUACXX_CAGATC_L002_R1
497097                                  300
100503874                                 0
100038431                                 0
19888                                     1
20671                                   182
27395                                   234
          MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1
497097                                   65
100503874                                 1
100038431                                 0
19888                                     0
20671                                    82
27395                                   337
          MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1
497097                                  237
100503874                                 1
100038431                                 0
19888                                     0
20671                                   105
27395                                   300
          MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1
497097                                  354
100503874                                 0
100038431                                 0
19888                                     0
20671                                    43
27395                                   290
          MCL1.DL_BC2CTUACXX_ATCACG_L002_R1
497097                                  287
100503874                                 4
100038431                                 0
19888                                     0
20671                                    82
27395                                   270
          MCL1.LA_BC2CTUACXX_GATCAG_L001_R1
497097                                    0
100503874                                 0
100038431                                 0
19888                                    10
20671                                    16
27395                                   560
          MCL1.LB_BC2CTUACXX_TGACCA_L001_R1
497097                                    0
100503874                                 0
100038431                                 0
19888                                     3
20671                                    25
27395                                   464
          MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1
497097                                    0
100503874                                 0
100038431                                 0
19888                                    10
20671                                    18
27395                                   489
          MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1
497097                                    0
100503874                                 0
100038431                                 0
19888                                     2
20671                                     8
27395                                   328
          MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1
497097                                    0
100503874                                 0
100038431                                 0
19888                                     0
20671                                     3
27395                                   307
          MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1
497097                                    0
100503874                                 0
100038431                                 0
19888                                     0
20671                                    10
27395                                   342
</code></pre>
<p>Now take a look at the column names</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(countdata)</code></pre></div>
<pre class="output"><code> [1] &quot;MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1&quot;
 [2] &quot;MCL1.DH_BC2CTUACXX_CAGATC_L002_R1&quot;
 [3] &quot;MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1&quot;
 [4] &quot;MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1&quot;
 [5] &quot;MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1&quot;
 [6] &quot;MCL1.DL_BC2CTUACXX_ATCACG_L002_R1&quot;
 [7] &quot;MCL1.LA_BC2CTUACXX_GATCAG_L001_R1&quot;
 [8] &quot;MCL1.LB_BC2CTUACXX_TGACCA_L001_R1&quot;
 [9] &quot;MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1&quot;
[10] &quot;MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1&quot;
[11] &quot;MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1&quot;
[12] &quot;MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1&quot;
</code></pre>
<p>These are the sample names which are pretty long so we’ll shorten these to contain only the relevant information about each sample. We will use the <code>substr</code> command to extract the first 7 characters and use these as the colnames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using substr, you extract the characters starting at position 1 and stopping at position 7 of the colnames</span>
<span class="kw">colnames</span>(countdata) &lt;-<span class="st"> </span><span class="kw">substr</span>(<span class="kw">colnames</span>(countdata),<span class="dt">start=</span><span class="dv">1</span>,<span class="dt">stop=</span><span class="dv">7</span>)</code></pre></div>
<p>Take a look at the output</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(countdata)</code></pre></div>
<pre class="output"><code>          MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB
497097        438     300      65     237     354     287       0       0
100503874       1       0       1       1       0       4       0       0
100038431       0       0       0       0       0       0       0       0
19888           1       1       0       0       0       0      10       3
20671         106     182      82     105      43      82      16      25
27395         309     234     337     300     290     270     560     464
          MCL1.LC MCL1.LD MCL1.LE MCL1.LF
497097          0       0       0       0
100503874       0       0       0       0
100038431       0       0       0       0
19888          10       2       0       0
20671          18       8       3      10
27395         489     328     307     342
</code></pre>
<p>Note that the column names are now the same as <code>SampleName</code> in the <code>sampleinfo</code> file. This is good because it means our sample information in <code>sampleinfo</code> is in the same order as the columns in <code>countdata</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">colnames</span>(countdata)==sampleinfo$SampleName)</code></pre></div>
<pre class="output"><code>
TRUE 
  12 
</code></pre>
<h2 id="filtering-to-remove-lowly-expressed-genes">Filtering to remove lowly expressed genes</h2>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>There are a few ways to filter out lowly expressed genes. When there are biological replicates in each group, in this case we have a sample size of 2 in each group, we favour filtering on a minimum counts per million threshold present in at least 2 samples. Two represents the smallest sample size for each group in our experiment. In this dataset, we choose to retain genes if they are expressed at a counts-per-million (CPM) above 0.5 in at least two samples.</p>
<p>We’ll use the <code>cpm</code> function from the <em>edgeR</em> library <span class="citation">(M D Robinson, McCarthy, and Smyth 2010)</span> to generate the CPM values and then filter. Note that by converting to CPMs we are normalising for the different sequencing depths for each sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Obtain CPMs</span>
myCPM &lt;-<span class="st"> </span><span class="kw">cpm</span>(countdata)
<span class="co"># Have a look at the output</span>
<span class="kw">head</span>(myCPM)</code></pre></div>
<pre class="output"><code>              MCL1.DG     MCL1.DH     MCL1.DI     MCL1.DJ   MCL1.DK
497097    18.85684388 13.77543859  2.69700983 10.45648006 16.442685
100503874  0.04305215  0.00000000  0.04149246  0.04412017  0.000000
100038431  0.00000000  0.00000000  0.00000000  0.00000000  0.000000
19888      0.04305215  0.04591813  0.00000000  0.00000000  0.000000
20671      4.56352843  8.35709941  3.40238163  4.63261775  1.997275
27395     13.30311589 10.74484210 13.98295863 13.23605071 13.469996
             MCL1.DL    MCL1.LA    MCL1.LB    MCL1.LC     MCL1.LD
497097    14.3389690  0.0000000  0.0000000  0.0000000  0.00000000
100503874  0.1998463  0.0000000  0.0000000  0.0000000  0.00000000
100038431  0.0000000  0.0000000  0.0000000  0.0000000  0.00000000
19888      0.0000000  0.4903857  0.1381969  0.4496078  0.09095771
20671      4.0968483  0.7846171  1.1516411  0.8092940  0.36383085
27395     13.4896224 27.4615975 21.3744588 21.9858214 14.91706476
             MCL1.LE    MCL1.LF
497097     0.0000000  0.0000000
100503874  0.0000000  0.0000000
100038431  0.0000000  0.0000000
19888      0.0000000  0.0000000
20671      0.1213404  0.4055595
27395     12.4171715 13.8701357
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Which values in myCPM are greater than 0.5?</span>
thresh &lt;-<span class="st"> </span>myCPM &gt;<span class="st"> </span><span class="fl">0.5</span>
<span class="co"># This produces a logical matrix with TRUEs and FALSEs</span>
<span class="kw">head</span>(thresh)</code></pre></div>
<pre class="output"><code>          MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB
497097       TRUE    TRUE    TRUE    TRUE    TRUE    TRUE   FALSE   FALSE
100503874   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE
100038431   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE
19888       FALSE   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE   FALSE
20671        TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE
27395        TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE
          MCL1.LC MCL1.LD MCL1.LE MCL1.LF
497097      FALSE   FALSE   FALSE   FALSE
100503874   FALSE   FALSE   FALSE   FALSE
100038431   FALSE   FALSE   FALSE   FALSE
19888       FALSE   FALSE   FALSE   FALSE
20671        TRUE   FALSE   FALSE   FALSE
27395        TRUE    TRUE    TRUE    TRUE
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Summary of how many TRUEs there are in each row</span>
<span class="co"># There are 11433 genes that have TRUEs in all 12 samples.</span>
<span class="kw">table</span>(<span class="kw">rowSums</span>(thresh))</code></pre></div>
<pre class="output"><code>
    0     1     2     3     4     5     6     7     8     9    10    11 
10857   518   544   307   346   307   652   323   547   343   579   423 
   12 
11433 
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we would like to keep genes that have at least 2 TRUES in each row of thresh</span>
keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(thresh) &gt;=<span class="st"> </span><span class="dv">2</span>
<span class="co"># Subset the rows of countdata to keep the more highly expressed genes</span>
counts.keep &lt;-<span class="st"> </span>countdata[keep,]
<span class="kw">summary</span>(keep)</code></pre></div>
<pre class="output"><code>   Mode   FALSE    TRUE    NA&#39;s 
logical   11375   15804       0 
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(counts.keep)</code></pre></div>
<pre class="output"><code>[1] 15804    12
</code></pre>
<p>A CPM of 0.5 is used as it corresponds to a count of 10-15 for the library sizes in this data set. If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample. A requirement for expression in two or more libraries is used as each group contains two replicates. This ensures that a gene will be retained if it is only expressed in one group. Smaller CPM thresholds are usually appropriate for larger libraries. As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.5. You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Let&#39;s have a look and see whether our threshold of 0.5 does indeed correspond to a count of about 10-15</span>
<span class="co"># We will look at the first sample</span>
<span class="kw">plot</span>(myCPM[,<span class="dv">1</span>],countdata[,<span class="dv">1</span>])</code></pre></div>
<p><img src="fig-06/unnamed-chunk-13-1.png" title="plot of chunk unnamed-chunk-13" alt="plot of chunk unnamed-chunk-13" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Let us limit the x and y-axis so we can actually look to see what is happening at the smaller counts</span>
<span class="kw">plot</span>(myCPM[,<span class="dv">1</span>],countdata[,<span class="dv">1</span>],<span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">50</span>),<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">3</span>))
<span class="co"># Add a vertical line at 0.5 CPM</span>
<span class="kw">abline</span>(<span class="dt">v=</span><span class="fl">0.5</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-13-2.png" title="plot of chunk unnamed-chunk-13" alt="plot of chunk unnamed-chunk-13" style="display: block; margin: auto;" /></p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Plot the counts-per-million versus counts for the second sample.</li>
<li>Add a vertical line at 0.5 and a horizontal line at 10.</li>
<li>Add the lines again, colouring them blue</li>
</ol>
<p>HINT: use the <code>col</code> parameter.</p>
</div>
</section>
<p><strong>Solution</strong> <img src="fig-06/unnamed-chunk-14-1.png" title="plot of chunk unnamed-chunk-14" alt="plot of chunk unnamed-chunk-14" style="display: block; margin: auto;" /> Note: When in doubt, a threshold of 1 CPM in at least <em>minimum</em> group sample size is a good rule of thumb.</p>
<h2 id="convert-counts-to-dgelist-object">Convert counts to DGEList object</h2>
<p>Next we’ll create a <code>DGEList</code> object. This is an object used by <em>edgeR</em> to store count data. It has a number of slots for storing various parameters about the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">DGEList</span>(counts.keep)
<span class="co"># have a look at y</span>
y</code></pre></div>
<pre class="output"><code>An object of class &quot;DGEList&quot;
$counts
       MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB
497097     438     300      65     237     354     287       0       0
20671      106     182      82     105      43      82      16      25
27395      309     234     337     300     290     270     560     464
18777      652     515     948     935     928     791     826     862
21399     1604    1495    1721    1317    1159    1066    1334    1258
       MCL1.LC MCL1.LD MCL1.LE MCL1.LF
497097       0       0       0       0
20671       18       8       3      10
27395      489     328     307     342
18777      668     646     544     581
21399     1068     926     508     500
15799 more rows ...

$samples
        group lib.size norm.factors
MCL1.DG     1 23218026            1
MCL1.DH     1 21768136            1
MCL1.DI     1 24091588            1
MCL1.DJ     1 22656713            1
MCL1.DK     1 21522033            1
7 more rows ...
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># See what slots are stored in y</span>
<span class="kw">names</span>(y)</code></pre></div>
<pre class="output"><code>[1] &quot;counts&quot;  &quot;samples&quot;
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Library size information is stored in the samples slot</span>
y$samples</code></pre></div>
<pre class="output"><code>        group lib.size norm.factors
MCL1.DG     1 23218026            1
MCL1.DH     1 21768136            1
MCL1.DI     1 24091588            1
MCL1.DJ     1 22656713            1
MCL1.DK     1 21522033            1
MCL1.DL     1 20008326            1
MCL1.LA     1 20384562            1
MCL1.LB     1 21698793            1
MCL1.LC     1 22235847            1
MCL1.LD     1 21982745            1
MCL1.LE     1 24719697            1
MCL1.LF     1 24652963            1
</code></pre>
<h2 id="quality-control">Quality control</h2>
<p>Now that we have got rid of the lowly expressed genes and have our counts stored in a <code>DGEList</code> object, we can look at a few different plots to check that the data is good quality, and that the samples are as we would expect.</p>
<h3 id="library-sizes-and-distribution-plots">Library sizes and distribution plots</h3>
<p>First, we can check how many reads we have for each sample in the <code>y</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y$samples$lib.size</code></pre></div>
<pre class="output"><code> [1] 23218026 21768136 24091588 22656713 21522033 20008326 20384562
 [8] 21698793 22235847 21982745 24719697 24652963
</code></pre>
<p>We can also plot the library sizes as a barplot to see whether there are any major discrepanies between the samples more easily.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The names argument tells the barplot to use the sample names on the x-axis</span>
<span class="co"># The las argument rotates the axis names</span>
<span class="kw">barplot</span>(y$samples$lib.size,<span class="dt">names=</span><span class="kw">colnames</span>(y),<span class="dt">las=</span><span class="dv">2</span>)
<span class="co"># Add a title to the plot</span>
<span class="kw">title</span>(<span class="st">&quot;Barplot of library sizes&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-17-1.png" title="plot of chunk unnamed-chunk-17" alt="plot of chunk unnamed-chunk-17" style="display: block; margin: auto;" /></p>
<p>Count data is not normally distributed, so if we want to examine the distributions of the raw counts we need to log the counts. Next we’ll use box plots to check the distribution of the read counts on the log2 scale. We can use the <code>cpm</code> function to get log2 counts per million, which are corrected for the different library sizes. The <code>cpm</code> function also adds a small offset to avoid taking log of zero.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get log2 counts per million</span>
logcounts &lt;-<span class="st"> </span><span class="kw">cpm</span>(y,<span class="dt">log=</span><span class="ot">TRUE</span>)
<span class="co"># Check distributions of samples using boxplots</span>
<span class="kw">boxplot</span>(logcounts, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Log2 counts per million&quot;</span>,<span class="dt">las=</span><span class="dv">2</span>)
<span class="co"># Let&#39;s add a blue horizontal line that corresponds to the median logCPM</span>
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(logcounts),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)
<span class="kw">title</span>(<span class="st">&quot;Boxplots of logCPMs (unnormalised)&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-18-1.png" title="plot of chunk unnamed-chunk-18" alt="plot of chunk unnamed-chunk-18" style="display: block; margin: auto;" /></p>
<p>From the boxplots we see that overall the density distributions of raw log-intensities are not identical but still not very different. If a sample is really far above or below the blue horizontal line we may need to investigate that sample further.</p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="discussion"><span class="glyphicon glyphicon-pencil"></span>Discussion</h2>
</div>
<div class="panel-body">
<p>Do any samples appear to be different compared to the others?</p>
</div>
</section>
<h3 id="multidimensional-scaling-plots">Multidimensional scaling plots</h3>
<p>By far, one of the most important plots we make when we analyse RNA-Seq data are MDSplots. An MDSplot is a visualisation of a principle components analysis, which determines the greatest sources of variation in the data. A principle components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code>plotMDS</code> function to create the MDS plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotMDS</span>(y)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-19-1.png" title="plot of chunk unnamed-chunk-19" alt="plot of chunk unnamed-chunk-19" style="display: block; margin: auto;" /></p>
<p>It is a bit difficult to see exactly what is going on with the default plot, although we do see samples grouping together in pairs. To make this plot more informative, we can colour the samples according to the grouping information. We can also change the labels, or instead of labels we can have points.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We specify the option to let us plot two plots side-by-sde</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="co"># Let&#39;s set up colour schemes for CellType</span>
<span class="co"># How many cell types and in what order are they stored?</span>
<span class="kw">levels</span>(sampleinfo$CellType)</code></pre></div>
<pre class="output"><code>[1] &quot;basal&quot;   &quot;luminal&quot;
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Let&#39;s choose purple for basal and orange for luminal
col.cell &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;orange&quot;</span>)[sampleinfo$CellType]
<span class="kw">data.frame</span>(sampleinfo$CellType,col.cell)</code></pre></div>
<pre class="output"><code>   sampleinfo.CellType col.cell
1              luminal   orange
2                basal   purple
3                basal   purple
4                basal   purple
5                basal   purple
6                basal   purple
7                basal   purple
8              luminal   orange
9              luminal   orange
10             luminal   orange
11             luminal   orange
12             luminal   orange
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Redo the MDS with cell type colouring</span>
<span class="kw">plotMDS</span>(y,<span class="dt">col=</span>col.cell)
<span class="co"># Let&#39;s add a legend to the plot so we know which colours correspond to which cell type</span>
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;orange&quot;</span>),<span class="dt">legend=</span><span class="kw">levels</span>(sampleinfo$CellType))
<span class="co"># Add a title</span>
<span class="kw">title</span>(<span class="st">&quot;Cell type&quot;</span>)

<span class="co"># Similarly for status</span>
<span class="kw">levels</span>(sampleinfo$Status)</code></pre></div>
<pre class="output"><code>[1] &quot;lactate&quot;  &quot;pregnant&quot; &quot;virgin&quot;  
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">col.status &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;dark green&quot;</span>)[sampleinfo$Status]
col.status</code></pre></div>
<pre class="output"><code> [1] &quot;dark green&quot; &quot;dark green&quot; &quot;red&quot;        &quot;red&quot;        &quot;blue&quot;      
 [6] &quot;blue&quot;       &quot;dark green&quot; &quot;dark green&quot; &quot;red&quot;        &quot;red&quot;       
[11] &quot;blue&quot;       &quot;blue&quot;      
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotMDS</span>(y,<span class="dt">col=</span>col.status)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;dark green&quot;</span>),<span class="dt">legend=</span><span class="kw">levels</span>(sampleinfo$Status),<span class="dt">cex=</span><span class="fl">0.8</span>)
<span class="kw">title</span>(<span class="st">&quot;Status&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-20-1.png" title="plot of chunk unnamed-chunk-20" alt="plot of chunk unnamed-chunk-20" style="display: block; margin: auto;" /></p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="discussion-1"><span class="glyphicon glyphicon-pencil"></span>Discussion</h2>
</div>
<div class="panel-body">
<p>Look at the MDS plot coloured by cell type. Is there something strange going on with the samples? Identify the two samples that don’t appear to be in the right place.</p>
</div>
</section>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># There is a sample info corrected file in your data directory</span>
<span class="co"># Old sampleinfo</span>
sampleinfo</code></pre></div>
<pre class="output"><code>                            FileName SampleName CellType   Status
1  MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1    MCL1.DG  luminal   virgin
2  MCL1.DH_BC2CTUACXX_CAGATC_L002_R1    MCL1.DH    basal   virgin
3  MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1    MCL1.DI    basal pregnant
4  MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1    MCL1.DJ    basal pregnant
5  MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1    MCL1.DK    basal  lactate
6  MCL1.DL_BC2CTUACXX_ATCACG_L002_R1    MCL1.DL    basal  lactate
7  MCL1.LA_BC2CTUACXX_GATCAG_L001_R1    MCL1.LA    basal   virgin
8  MCL1.LB_BC2CTUACXX_TGACCA_L001_R1    MCL1.LB  luminal   virgin
9  MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1    MCL1.LC  luminal pregnant
10 MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1    MCL1.LD  luminal pregnant
11 MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1    MCL1.LE  luminal  lactate
12 MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1    MCL1.LF  luminal  lactate
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># I&#39;m going to write over the sampleinfo object with the corrected sample info</span>
sampleinfo &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/SampleInfo_Corrected.txt&quot;</span>)
sampleinfo</code></pre></div>
<pre class="output"><code>                            FileName SampleName CellType   Status
1  MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1    MCL1.DG    basal   virgin
2  MCL1.DH_BC2CTUACXX_CAGATC_L002_R1    MCL1.DH    basal   virgin
3  MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1    MCL1.DI    basal pregnant
4  MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1    MCL1.DJ    basal pregnant
5  MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1    MCL1.DK    basal  lactate
6  MCL1.DL_BC2CTUACXX_ATCACG_L002_R1    MCL1.DL    basal  lactate
7  MCL1.LA_BC2CTUACXX_GATCAG_L001_R1    MCL1.LA  luminal   virgin
8  MCL1.LB_BC2CTUACXX_TGACCA_L001_R1    MCL1.LB  luminal   virgin
9  MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1    MCL1.LC  luminal pregnant
10 MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1    MCL1.LD  luminal pregnant
11 MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1    MCL1.LE  luminal  lactate
12 MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1    MCL1.LF  luminal  lactate
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Redo the MDSplot with corrected information</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
col.cell &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;orange&quot;</span>)[sampleinfo$CellType]
col.status &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;dark green&quot;</span>)[sampleinfo$Status]
<span class="kw">plotMDS</span>(y,<span class="dt">col=</span>col.cell)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;orange&quot;</span>),<span class="dt">legend=</span><span class="kw">levels</span>(sampleinfo$CellType))
<span class="kw">title</span>(<span class="st">&quot;Cell type&quot;</span>)
<span class="kw">plotMDS</span>(y,<span class="dt">col=</span>col.status)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;dark green&quot;</span>),<span class="dt">legend=</span><span class="kw">levels</span>(sampleinfo$Status),<span class="dt">cex=</span><span class="fl">0.8</span>)
<span class="kw">title</span>(<span class="st">&quot;Status&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-22-1.png" title="plot of chunk unnamed-chunk-22" alt="plot of chunk unnamed-chunk-22" style="display: block; margin: auto;" /></p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="discussion-2"><span class="glyphicon glyphicon-pencil"></span>Discussion</h2>
</div>
<div class="panel-body">
<p>What is the greatest source of variation in the data (i.e. what does dimension 1 represent)? What is the second greatest source of variation in the data?</p>
</div>
</section>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-1"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Redo the plots choosing your own colours.</li>
<li>Change the plotting character to a symbol instead of the column names HINT: use <code>pch</code> argument. Try <code>pch=16</code> and see what happens.</li>
<li>Change the plotting characters such that basal samples have the value <code>1</code> and luminal samples have the value <code>4</code>. Colour by status (lactate, pregnant, virgin)</li>
</ol>
</div>
</section>
<p><strong>Solution</strong></p>
<pre class="output"><code>[1] &quot;#1B9E77&quot; &quot;#D95F02&quot; &quot;#7570B3&quot;
</code></pre>
<p><img src="fig-06/unnamed-chunk-23-1.png" title="plot of chunk unnamed-chunk-23" alt="plot of chunk unnamed-chunk-23" style="display: block; margin: auto;" /></p>
<p>The distance between each pair of samples in the MDS plot is calculated as the leading fold change, defined as the root-mean-square of the largest 500 log2-fold changes between that pair of samples. Replicate samples from the same group cluster together in the plot, while samples from different groups form separate clusters. This indicates that the differences between groups are larger than those within groups, i.e., differential expression is greater than the variance and can be detected. In the MDS plot, the distance between basal samples on the left and luminal cells on the right is about 6 units, corresponding to a leading fold change of about 64-fold (2^6 = 64) between basal and luminal. The expression differences between virgin, pregnant and lactating are greater for luminal cells than for basal.</p>
<p>Notes</p>
<ul>
<li>The MDS plot can be simply generated with <code>plotMDS(y)</code>. The additional code is purely for aesthetics, to improve the visualization of the groups.</li>
<li>Clustering in the MDS plot can be used to motivate changes to the design matrix in light of potential batch effects. For example, imagine that the first replicate of each group was prepared at a separate time from the second replicate. If the MDS plot showed separation of samples by time, it might be worthwhile including time in the down stream analysis to account for the time-based effect.</li>
</ul>
<p><code>plotMDS</code> plots the first two dimensions as a default, however you can plot higher dimensions using the <code>dim</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Dimension 3 appears to separate pregnant samples from the rest. Dim4?</span>
<span class="kw">plotMDS</span>(y,<span class="dt">dim=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>),<span class="dt">col=</span>col.status,<span class="dt">pch=</span>char.celltype,<span class="dt">cex=</span><span class="dv">2</span>)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>,<span class="dt">legend=</span><span class="kw">levels</span>(sampleinfo$Status),<span class="dt">col=</span>cols,<span class="dt">pch=</span><span class="dv">16</span>)
<span class="kw">legend</span>(<span class="st">&quot;bottomright&quot;</span>,<span class="dt">legend=</span><span class="kw">levels</span>(sampleinfo$CellType),<span class="dt">pch=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>))</code></pre></div>
<p><img src="fig-06/unnamed-chunk-24-1.png" title="plot of chunk unnamed-chunk-24" alt="plot of chunk unnamed-chunk-24" style="display: block; margin: auto;" /></p>
<p>Another alternative is to generate an interactive MDS plot using the <em>Glimma</em> package. This allows the user to interactively explore the different dimensions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">labels &lt;-<span class="st"> </span><span class="kw">paste</span>(sampleinfo$SampleName, sampleinfo$CellType, sampleinfo$Status)
group &lt;-<span class="st"> </span><span class="kw">paste</span>(sampleinfo$CellType,sampleinfo$Status,<span class="dt">sep=</span><span class="st">&quot;.&quot;</span>)
group &lt;-<span class="st"> </span><span class="kw">factor</span>(group)
<span class="kw">glMDSPlot</span>(y, <span class="dt">labels=</span>labels, <span class="dt">groups=</span>group, <span class="dt">folder=</span><span class="st">&quot;mds&quot;</span>)</code></pre></div>
<p><em>Glimma</em> was created to make interactive versions of some of the popular plots from the <em>limma</em> package. At present it can be used to obtain MDS plots and mean-difference (MD) plots, which will be covered later. The output of <code>glMDSPlot</code> is an html page (/mds/MDS-Plot.html) that shows the MDS plot on the left, and the amount of variation explained by each dimension in a barplot on the right. The user can hover over points to find out sample information, and switch between successive dimensions in the MDS plot by clicking on the bars in the barplot. The default MDS plots shows dimensions 1 and 2.</p>
<h3 id="hierarchical-clustering-with-heatmaps">Hierarchical clustering with heatmaps</h3>
<p>An alternative to <code>plotMDS</code> for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. We can do this using the <code>heatmap.2</code> function from the <em>gplots</em> package. In this example <code>heatmap.2</code> calculates a matrix of euclidean distances from the logCPM (<code>logcounts</code> object) for the 500 most variable genes. (Note this has more complicated code than plotting principle components using <code>plotMDS</code>.)</p>
<p>The <em>RColorBrewer</em> package has nicer colour schemes, accessed using the <code>brewer.pal</code> function. “RdYlBu” is a common choice, and “Spectral” is also nice.</p>
<p>Note:The <code>png</code> function will create a png file to save the plots created straight after, and will close this file when <code>dev.off()</code> is called. To see your plots interactively, simply omit those two lines.</p>
<p>Let’s select data for the 500 most variable genes and plot the heatmap</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We estimate the variance for each row in the logcounts matrix</span>
var_genes &lt;-<span class="st"> </span><span class="kw">apply</span>(logcounts, <span class="dv">1</span>, var)
<span class="kw">head</span>(var_genes)</code></pre></div>
<pre class="output"><code>    497097      20671      27395      18777      21399      58175 
27.4382296  3.3141188  0.1596456  0.1313426  0.3947091  6.2321017 
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get the gene names for the top 500 most variable genes</span>
select_var &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(var_genes, <span class="dt">decreasing=</span><span class="ot">TRUE</span>))[<span class="dv">1</span>:<span class="dv">500</span>]
<span class="kw">head</span>(select_var)</code></pre></div>
<pre class="output"><code>[1] &quot;16846&quot;  &quot;72902&quot;  &quot;56747&quot;  &quot;497097&quot; &quot;24117&quot;  &quot;109205&quot;
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Subset logcounts matrix</span>
highly_variable_lcpm &lt;-<span class="st"> </span>logcounts[select_var,]
<span class="kw">dim</span>(highly_variable_lcpm)</code></pre></div>
<pre class="output"><code>[1] 500  12
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(highly_variable_lcpm)</code></pre></div>
<pre class="output"><code>         MCL1.DG  MCL1.DH  MCL1.DI  MCL1.DJ  MCL1.DK  MCL1.DL   MCL1.LA
16846  0.4290261 1.394877 4.025241 6.068229 5.923462 5.467369 -6.486175
72902  1.4460253 2.569794 6.084854 5.183325 5.216547 4.629406 -6.486175
56747  3.3347872 2.975754 5.071799 5.016300 5.599113 6.048402 -6.486175
497097 4.2384664 3.785840 1.437863 3.388414 4.040841 3.843500 -6.486175
24117  7.8373931 7.899107 8.634338 8.582153 6.703626 6.776308 -1.710778
109205 5.5238723 5.717292 6.197664 6.203120 6.526628 6.529395 -6.486175
          MCL1.LB   MCL1.LC     MCL1.LD   MCL1.LE   MCL1.LF
16846  -4.1268214 -4.155152 -6.48617500 -6.486175 -6.486175
72902  -6.4861750 -3.306148 -4.14190372 -6.486175 -6.486175
56747  -6.4861750 -3.306148 -6.48617500 -4.276248 -4.273190
497097 -6.4861750 -6.486175 -6.48617500 -6.486175 -6.486175
24117  -0.7124885 -0.366455  0.01712352 -4.276248 -4.273190
109205 -1.7975234 -2.388018 -3.29144547 -4.276248 -2.527764
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Get some nicer colours
mypalette &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">11</span>,<span class="st">&quot;RdYlBu&quot;</span>)
morecols &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(mypalette)
<span class="co"># Set up colour vector for celltype variable</span>
col.cell &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;purple&quot;</span>,<span class="st">&quot;orange&quot;</span>)[sampleinfo$CellType]

<span class="co"># Plot the heatmap</span>
<span class="kw">heatmap.2</span>(highly_variable_lcpm,<span class="dt">col=</span><span class="kw">rev</span>(<span class="kw">morecols</span>(<span class="dv">50</span>)),<span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">main=</span><span class="st">&quot;Top 500 most variable genes across samples&quot;</span>,<span class="dt">ColSideColors=</span>col.cell,<span class="dt">scale=</span><span class="st">&quot;row&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-26-1.png" title="plot of chunk unnamed-chunk-26" alt="plot of chunk unnamed-chunk-26" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Save the heatmap</span>
<span class="kw">png</span>(<span class="dt">file=</span><span class="st">&quot;High_var_genes.heatmap.png&quot;</span>)
<span class="kw">heatmap.2</span>(highly_variable_lcpm,<span class="dt">col=</span><span class="kw">rev</span>(<span class="kw">morecols</span>(<span class="dv">50</span>)),<span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">main=</span><span class="st">&quot;Top 500 most variable genes across samples&quot;</span>,<span class="dt">ColSideColors=</span>col.cell,<span class="dt">scale=</span><span class="st">&quot;row&quot;</span>)
<span class="kw">dev.off</span>()</code></pre></div>
<pre class="output"><code>quartz_off_screen 
                2 
</code></pre>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-2"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Change the colour scheme to “PiYG” and redo the heatmap. Try <code>?RColorBrewer</code> and see what other colour schemes are available.</li>
<li>Change the sample names to <code>group</code> using the <code>labCol</code> argument</li>
<li>Redo the heatmap using the top 500 LEAST variable genes.</li>
</ol>
</div>
</section>
<p><strong>Solution</strong> <img src="fig-06/unnamed-chunk-27-1.png" title="plot of chunk unnamed-chunk-27" alt="plot of chunk unnamed-chunk-27" style="display: block; margin: auto;" /></p>
<pre class="output"><code>[1] &quot;320714&quot; &quot;93684&quot;  &quot;245638&quot; &quot;26433&quot;  &quot;93747&quot;  &quot;236792&quot;
</code></pre>
<p><img src="fig-06/unnamed-chunk-27-2.png" title="plot of chunk unnamed-chunk-27" alt="plot of chunk unnamed-chunk-27" style="display: block; margin: auto;" /></p>
<hr />
<h2 id="normalisation-for-composition-bias">Normalisation for composition bias</h2>
<p>TMM normalization is performed to eliminate composition biases between libraries <span class="citation">(Mark D Robinson and Oshlack 2010)</span>. This generates a set of normalization factors, where the product of these factors and the library sizes defines the effective library size. The <code>calcNormFactors</code> function calculates the normalization factors between libraries. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Apply normalisation to DGEList object</span>
y &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(y)</code></pre></div>
<p>This will update the normalisation factors in the <code>DGEList</code> object (their default values are 1). Take a look at the normalisation factors for these samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y$samples</code></pre></div>
<pre class="output"><code>        group lib.size norm.factors
MCL1.DG     1 23218026    1.2368993
MCL1.DH     1 21768136    1.2139485
MCL1.DI     1 24091588    1.1255640
MCL1.DJ     1 22656713    1.0698261
MCL1.DK     1 21522033    1.0359212
MCL1.DL     1 20008326    1.0872153
MCL1.LA     1 20384562    1.3684449
MCL1.LB     1 21698793    1.3653200
MCL1.LC     1 22235847    1.0047431
MCL1.LD     1 21982745    0.9232822
MCL1.LE     1 24719697    0.5291015
MCL1.LF     1 24652963    0.5354877
</code></pre>
<p>The normalization factors multiply to unity across all libraries. A normalization factor below one indicates that the library size will be scaled down, as there is more suppression (i.e., composition bias) in that library relative to the other libraries. This is also equivalent to scaling the counts upwards in that sample. Conversely, a factor above one scales up the library size and is equivalent to downscaling the counts.</p>
<p>The last two samples have much smaller normalisation factors, and MCL1.LA and MCL1.LB have the largest. If we plot mean difference plots using the <code>plotMD</code> function for these samples, we should be able to see the composition bias problem. We will use the <code>logcounts</code>, which have been normalised for library size, but not for composition bias.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMD</span>(logcounts,<span class="dt">column =</span> <span class="dv">7</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)
<span class="kw">plotMD</span>(logcounts,<span class="dt">column =</span> <span class="dv">11</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-30-1.png" title="plot of chunk unnamed-chunk-30" alt="plot of chunk unnamed-chunk-30" style="display: block; margin: auto;" /></p>
<p>The mean-difference plots show average expression (mean: x-axis) against log-fold-changes (difference: y-axis). Because our <code>DGEList</code> object contains the normalisation factors, if we redo these plots using <code>y</code>, we should see the composition bias problem has been solved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMD</span>(y,<span class="dt">column =</span> <span class="dv">7</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)
<span class="kw">plotMD</span>(y,<span class="dt">column =</span> <span class="dv">11</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-31-1.png" title="plot of chunk unnamed-chunk-31" alt="plot of chunk unnamed-chunk-31" style="display: block; margin: auto;" /></p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-3"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<p>Plot the biased and unbiased MD plots side by side for the same sample to see the before and after TMM normalisation effect.</p>
</div>
</section>
<p><strong>Solution</strong> <img src="fig-06/unnamed-chunk-32-1.png" title="plot of chunk unnamed-chunk-32" alt="plot of chunk unnamed-chunk-32" style="display: block; margin: auto;" /></p>
<p><strong>We need to save a few data objects to use for Day 2 so we don’t have to rerun everything</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># save(group,y,logcounts,sampleinfo,file=&quot;day1objects.Rdata&quot;)</span></code></pre></div>
<h2 id="differential-expression-with-limma-voom">Differential expression with limma-voom</h2>
<p>Now that we are happy that we have normalised the data and that the quality looks good, we can continue to testing for differentially expressed genes. There are a number of packages to analyse RNA-Seq data. The <em>limma</em> package <span class="citation">(Ritchie et al. 2015)</span> (since version 3.16.0) offers the <code>voom</code> function, which transforms the read counts into logCPMs while taking into account the mean-variance relationship in the data <span class="citation">(C. W. Law et al. 2014)</span>. After vooming, users can apply a linear model to the voom transformed data to test for differentially expressed genes, using standard <em>limma</em> commands.</p>
<p><strong>Load the objects into the workspace that we created yesterday</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load(&quot;day1objects.Rdata&quot;)</span>
<span class="co"># objects()</span></code></pre></div>
<h3 id="create-the-design-matrix">Create the design matrix</h3>
<p>First we need to create a design matrix for the groups (see the excellent <a href="https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma user guide</a> for more information on design matrices). There are many different ways to set up your design matrix, and it is dictated by what comparisons you would like to test. We will follow the set-up from pg 43 of the limma vignette (“Interaction models: 2X2 factorial designs”).</p>
<p>In this analysis let’s assume that we will be testing differences in status in the different cell types separately. For example, we want to know which genes are differentially expressed between pregnant and lactating in basal cells only. We have previously codedthe <code>group</code> variable, which is a concatenation of cell type and status. Coding the cell type and status in this way allows us to be flexible in specifying which comparisons we are interested in.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at group variable again</span>
group</code></pre></div>
<pre class="output"><code> [1] basal.virgin     basal.virgin     basal.pregnant   basal.pregnant  
 [5] basal.lactate    basal.lactate    luminal.virgin   luminal.virgin  
 [9] luminal.pregnant luminal.pregnant luminal.lactate  luminal.lactate 
6 Levels: basal.lactate basal.pregnant basal.virgin ... luminal.virgin
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Specify a design matrix without an intercept term</span>
design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(~<span class="st"> </span><span class="dv">0</span> +<span class="st"> </span>group)
design</code></pre></div>
<pre class="output"><code>   groupbasal.lactate groupbasal.pregnant groupbasal.virgin
1                   0                   0                 1
2                   0                   0                 1
3                   0                   1                 0
4                   0                   1                 0
5                   1                   0                 0
6                   1                   0                 0
7                   0                   0                 0
8                   0                   0                 0
9                   0                   0                 0
10                  0                   0                 0
11                  0                   0                 0
12                  0                   0                 0
   groupluminal.lactate groupluminal.pregnant groupluminal.virgin
1                     0                     0                   0
2                     0                     0                   0
3                     0                     0                   0
4                     0                     0                   0
5                     0                     0                   0
6                     0                     0                   0
7                     0                     0                   1
8                     0                     0                   1
9                     0                     1                   0
10                    0                     1                   0
11                    1                     0                   0
12                    1                     0                   0
attr(,&quot;assign&quot;)
[1] 1 1 1 1 1 1
attr(,&quot;contrasts&quot;)
attr(,&quot;contrasts&quot;)$group
[1] &quot;contr.treatment&quot;
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Make the column names of the design matrix a bit nicer
<span class="kw">colnames</span>(design) &lt;-<span class="st"> </span><span class="kw">levels</span>(group)
design</code></pre></div>
<pre class="output"><code>   basal.lactate basal.pregnant basal.virgin luminal.lactate
1              0              0            1               0
2              0              0            1               0
3              0              1            0               0
4              0              1            0               0
5              1              0            0               0
6              1              0            0               0
7              0              0            0               0
8              0              0            0               0
9              0              0            0               0
10             0              0            0               0
11             0              0            0               1
12             0              0            0               1
   luminal.pregnant luminal.virgin
1                 0              0
2                 0              0
3                 0              0
4                 0              0
5                 0              0
6                 0              0
7                 0              1
8                 0              1
9                 1              0
10                1              0
11                0              0
12                0              0
attr(,&quot;assign&quot;)
[1] 1 1 1 1 1 1
attr(,&quot;contrasts&quot;)
attr(,&quot;contrasts&quot;)$group
[1] &quot;contr.treatment&quot;
</code></pre>
<p>Each column of the design matrix tells us which samples correspond to each group. The samples which come from basal cells from a lactating mouse correspond to columns 5 and 6 in the counts matrix, i.e. the samples which have 1s.</p>
<h3 id="voom-transform-the-data">Voom transform the data</h3>
<p>Once we have our design matrix ready to go, we can perform our voom transformation. Voom will automatically adjust the library sizes using the <code>norm.factors</code> already calculated. The voom transformation uses the experiment design matrix, and produces an <code>EList</code> object. We can add <code>plot=TRUE</code> to generate a plot of the mean-variance trend. This plot can also tell us if there are any genes that look really variable in our data, and if we’ve filtered the low counts adequately.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
v &lt;-<span class="st"> </span><span class="kw">voom</span>(y,design,<span class="dt">plot =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-36-1.png" title="plot of chunk unnamed-chunk-36" alt="plot of chunk unnamed-chunk-36" style="display: block; margin: auto;" /></p>
<p>The voom normalised log2 counts can be found in v$E. Take a look at what is in the voom object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">v</code></pre></div>
<pre class="output"><code>An object of class &quot;EList&quot;
$targets
        group lib.size norm.factors
MCL1.DG     1 28718360     1.236899
MCL1.DH     1 26425396     1.213949
MCL1.DI     1 27116624     1.125564
MCL1.DJ     1 24238743     1.069826
MCL1.DK     1 22295130     1.035921
7 more rows ...

$E
        MCL1.DG  MCL1.DH  MCL1.DI  MCL1.DJ   MCL1.DK  MCL1.DL    MCL1.LA
497097 3.932532 3.507368 1.272317 3.292541 3.9909851 3.724252 -5.8019424
20671  1.890808 2.787899 1.605217 2.121856 0.9642868 1.923156 -0.7575483
27395  3.429894 3.149591 3.637638 3.631978 3.7037376 3.636318  4.3286281
18777  4.505933 4.285975 5.128398 5.270351 5.3801014 5.185279  4.8889286
21399  5.804007 5.822559 5.988345 5.764344 5.7006304 5.615502  5.5801412
          MCL1.LB    MCL1.LC   MCL1.LD   MCL1.LE    MCL1.LF
497097 -5.8887821 -5.4816421 -5.343143 -4.709206 -4.7226146
20671  -0.2163568 -0.2721888 -1.255680 -1.901851 -0.3302972
27395   3.9707527  4.4535229  4.016606  4.555237  4.6973455
18777   4.8635986  4.9031416  4.993363  5.379583  5.4610208
21399   5.4087074  5.5797290  5.512504  5.280898  5.2446116
15799 more rows ...

$weights
         [,1]     [,2]     [,3]     [,4]     [,5]     [,6]       [,7]
[1,] 42.85032 40.59638 19.78119 18.11020 38.62541 37.98661  0.7907909
[2,] 21.34610 19.99539 15.74020 14.40027 10.70068 10.49511  4.5401043
[3,] 35.05544 33.04583 39.72941 36.82081 35.31518 34.70612 50.4586968
[4,] 56.26175 53.86441 69.16366 66.65108 66.03797 65.46069 64.6692219
[5,] 76.92857 75.98600 76.79528 75.45282 71.62644 71.16558 73.5980084
           [,8]       [,9]      [,10]      [,11]      [,12]
[1,]  0.7907909  0.7907909  0.7907909  0.7907909  0.7907909
[2,]  4.7399033  3.4277768  3.2224692  2.1314506  2.1428929
[3,] 52.1833006 45.7806579 43.1302387 38.6112610 38.8559918
[4,] 66.0984998 60.0161723 57.3174922 54.0693981 54.3410792
[5,] 74.5613544 70.1642210 68.1238307 50.8570652 51.1210163
15799 more rows ...

$design
  basal.lactate basal.pregnant basal.virgin luminal.lactate
1             0              0            1               0
2             0              0            1               0
3             0              1            0               0
4             0              1            0               0
5             1              0            0               0
  luminal.pregnant luminal.virgin
1                0              0
2                0              0
3                0              0
4                0              0
5                0              0
7 more rows ...
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># What is contained in this object?</span>
<span class="kw">names</span>(v)</code></pre></div>
<pre class="output"><code>[1] &quot;targets&quot; &quot;E&quot;       &quot;weights&quot; &quot;design&quot; 
</code></pre>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-4"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>What is in the <code>targets</code> slot of <code>v</code> and what does it correspond to in <code>y</code>?</li>
<li>What are the dimensions of the <code>weights</code> slot in <code>v</code>?</li>
</ol>
</div>
</section>
<p>We can repeat the box plots for the normalised data to compare to before normalisation. The expression values in <code>v$E</code> are already log2 values so we don’t need to log-transform.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">boxplot</span>(logcounts, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Log2 counts per million&quot;</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">main=</span><span class="st">&quot;Unnormalised logCPM&quot;</span>)
## Let&#39;s add a blue horizontal line that corresponds to the median logCPM
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(logcounts),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)
<span class="kw">boxplot</span>(v$E, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Log2 counts per million&quot;</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">main=</span><span class="st">&quot;Voom transformed logCPM&quot;</span>)
## Let&#39;s add a blue horizontal line that corresponds to the median logCPM
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(v$E),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-38-1.png" title="plot of chunk unnamed-chunk-38" alt="plot of chunk unnamed-chunk-38" style="display: block; margin: auto;" /></p>
<p>Compare these box plots to the box plots we generated before performing the normalisation. Can you see any differences?</p>
<h3 id="testing-for-differential-expression">Testing for differential expression</h3>
<p>Now that we have the voom transformed data we can use <em>limma</em> to test for differential expression. First we fit a linear model for each gene using the <code>lmFit</code> function in <em>limma</em>. <code>lmFit</code> needs the voom object and the design matrix that we have already specified, which is stored within the voom object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit the linear model</span>
fit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(v)
<span class="kw">names</span>(fit)</code></pre></div>
<pre class="output"><code> [1] &quot;coefficients&quot;     &quot;stdev.unscaled&quot;   &quot;sigma&quot;           
 [4] &quot;df.residual&quot;      &quot;cov.coefficients&quot; &quot;pivot&quot;           
 [7] &quot;rank&quot;             &quot;Amean&quot;            &quot;method&quot;          
[10] &quot;design&quot;          
</code></pre>
<p><code>lmFit</code> estimates group means according to the design matrix, as well as gene-wise variances. There are a number of items stored in the <code>fit</code> object, most of which are specific to the statistical testing, and we won’t be discussing these in detail today.</p>
<p>Since we are interested in differences between groups, we need to specify which comparisons we want to test. The comparison of interest can be specified using the <code>makeContrasts</code> function. Here, we are interested in knowing which genes are differentially expressed between the pregnant and lactating group in the basal cells. This is done by defining the null hypothesis as basal.pregnant - basal.lactate = 0 for each gene. Note that the group names must exactly match the column names of the design matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cont.matrix &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(<span class="dt">B.PregVsLac=</span>basal.pregnant -<span class="st"> </span>basal.lactate,<span class="dt">levels=</span>design)</code></pre></div>
<p>Take a look at the contrast matrix. The contrast matrix tells <em>limma</em> which columns of the design matrix we are interested in testing our comparison. Note that here we have specified only one comparison to test, but we can specify as many as we want in one go.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cont.matrix</code></pre></div>
<pre class="output"><code>                  Contrasts
Levels             B.PregVsLac
  basal.lactate             -1
  basal.pregnant             1
  basal.virgin               0
  luminal.lactate            0
  luminal.pregnant           0
  luminal.virgin             0
</code></pre>
<p>Now we can apply the contrasts matrix to the <code>fit</code> object to get the statistics and estimated parameters of our comparison that we are interested in. Here we call the <code>contrasts.fit</code> function in <em>limma</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.cont &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(fit, cont.matrix)</code></pre></div>
<p>The final step is to call the <code>eBayes</code> function, which performs empirical Bayes shrinkage on the variances, and estimates moderated t-statistics and the associated p-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.cont &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit.cont)</code></pre></div>
<p>Check the dimensions of the fit object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(fit.cont)</code></pre></div>
<pre class="output"><code>[1] 15804     1
</code></pre>
<p>We can use the <em>limma</em> <code>decideTests</code> function to generate a quick summary of DE genes for the contrasts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">summa.fit &lt;-<span class="st"> </span><span class="kw">decideTests</span>(fit.cont)
<span class="kw">summary</span>(summa.fit)</code></pre></div>
<pre class="output"><code>   B.PregVsLac
-1        2635
0        10464
1         2705
</code></pre>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-5"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Add another contrast to the contrasts matrix: <code>L.PregVsLac = luminal.pregnant - luminal.lactate</code> and re-run the code above. You should have two comparisons in <code>fit.cont</code> now.</li>
<li>Check out the <code>vennDiagram</code> function (HINT: type <code>?vennDiagram</code>). Can you show the overlap of differentially expressed genes between the two comparisons? How many genes are commonly differentially expressed?</li>
</ol>
</div>
</section>
<p><strong>Solution</strong></p>
<pre class="output"><code>   B.PregVsLac L.PregVsLac
-1        2635        3686
0        10464        8451
1         2705        3667
</code></pre>
<p><img src="fig-06/unnamed-chunk-46-1.png" title="plot of chunk unnamed-chunk-46" alt="plot of chunk unnamed-chunk-46" style="display: block; margin: auto;" /></p>
<p>The <em>limma</em> <code>topTable</code> function summarises the output in a table format. Significant DE genes for a particular comparison can be identified by selecting genes with a p-value smaller than a chosen cut-off value and/or a fold change greater than a chosen value in this table. By default the table will be sorted by the B statistic, which is the log-odds of differential expression. Usually the B statistic and p-value ranking will be the same, but this is not always the case. We will explicitly rank by p-value, which we can specify with the <code>sort.by</code> argument.</p>
<p>The <code>topTable</code> command will always output the top 10 genes by default, even if they are not statistically significant. We can specify the coefficient we are interested in by the name we used in the contrast matrix (“B.PregVsLac”), or by the column number.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit.cont,<span class="dt">coef=</span><span class="st">&quot;B.PregVsLac&quot;</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>           logFC  AveExpr         t      P.Value    adj.P.Val        B
24117   1.819943 2.975545  20.10780 1.063770e-10 1.016240e-06 14.96977
381290 -2.143885 3.944066 -19.07495 1.982934e-10 1.016240e-06 14.39556
78896   2.807548 3.036519  18.54773 2.758828e-10 1.016240e-06 14.07416
226101 -2.329744 6.223525 -18.26861 3.297667e-10 1.016240e-06 13.85802
16012  -2.896115 1.978449 -18.21525 3.413066e-10 1.016240e-06 13.46984
231830  2.253400 4.760597  18.02627 3.858161e-10 1.016240e-06 13.67600
16669  -2.312721 8.741892 -17.07937 7.264548e-10 1.640127e-06 13.17357
55987  -1.515469 2.834512 -16.64333 9.829870e-10 1.718703e-06 12.88782
231991 -2.598105 4.275930 -16.53634 1.059885e-09 1.718703e-06 12.66020
14620   3.600094 3.525281  16.46627 1.113755e-09 1.718703e-06 12.54071
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## This will give the same output
<span class="kw">topTable</span>(fit.cont,<span class="dt">coef=</span><span class="dv">1</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>           logFC  AveExpr         t      P.Value    adj.P.Val        B
24117   1.819943 2.975545  20.10780 1.063770e-10 1.016240e-06 14.96977
381290 -2.143885 3.944066 -19.07495 1.982934e-10 1.016240e-06 14.39556
78896   2.807548 3.036519  18.54773 2.758828e-10 1.016240e-06 14.07416
226101 -2.329744 6.223525 -18.26861 3.297667e-10 1.016240e-06 13.85802
16012  -2.896115 1.978449 -18.21525 3.413066e-10 1.016240e-06 13.46984
231830  2.253400 4.760597  18.02627 3.858161e-10 1.016240e-06 13.67600
16669  -2.312721 8.741892 -17.07937 7.264548e-10 1.640127e-06 13.17357
55987  -1.515469 2.834512 -16.64333 9.829870e-10 1.718703e-06 12.88782
231991 -2.598105 4.275930 -16.53634 1.059885e-09 1.718703e-06 12.66020
14620   3.600094 3.525281  16.46627 1.113755e-09 1.718703e-06 12.54071
</code></pre>
<h3 id="adding-annotation-and-saving-the-results">Adding annotation and saving the results</h3>
<p>We have a list of significantly differentially expressed genes, but the only annotation we can see is the Entrez Gene ID, which is not very informative. We would like to add some annotation information. There are a number of ways to do this. We will demonstrate how to do this using the <em>org.Mm.eg.db</em> package.</p>
<p>First we need to decide what information we want. In order to see what we can extract we can run the <code>columns</code> function on the annotation database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">columns</span>(org.Mm.eg.db)</code></pre></div>
<pre class="output"><code> [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot; 
 [5] &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;    
 [9] &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;       
[13] &quot;IPI&quot;          &quot;MGI&quot;          &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot; 
[17] &quot;PATH&quot;         &quot;PFAM&quot;         &quot;PMID&quot;         &quot;PROSITE&quot;     
[21] &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UNIGENE&quot;      &quot;UNIPROT&quot;     
</code></pre>
<p>We definitely want gene symbols and perhaps the full gene name. Let’s build up our annotation information in a separate data frame using the <code>select</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ann &lt;-<span class="st"> </span><span class="kw">select</span>(org.Mm.eg.db,<span class="dt">keys=</span><span class="kw">rownames</span>(fit.cont),<span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;ENTREZID&quot;</span>,<span class="st">&quot;SYMBOL&quot;</span>,<span class="st">&quot;GENENAME&quot;</span>))</code></pre></div>
<pre class="output"><code>&#39;select()&#39; returned 1:1 mapping between keys and columns
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Have a look at the annotation</span>
<span class="kw">head</span>(ann)</code></pre></div>
<pre class="output"><code>  ENTREZID SYMBOL                                  GENENAME
1   497097   Xkr4         X-linked Kx blood group related 4
2    20671  Sox17     SRY (sex determining region Y)-box 17
3    27395 Mrpl15       mitochondrial ribosomal protein L15
4    18777 Lypla1                       lysophospholipase 1
5    21399  Tcea1 transcription elongation factor A (SII) 1
6    58175  Rgs20       regulator of G-protein signaling 20
</code></pre>
<p>Let’s double check that the <code>ENTREZID</code> column matches exactly to our <code>fit.cont</code> rownames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ann$ENTREZID==<span class="kw">rownames</span>(fit.cont))</code></pre></div>
<pre class="output"><code>
 TRUE 
15804 
</code></pre>
<p>We can slot in the annotation information into the <code>genes</code> slot of <code>fit.cont</code>. (Please note that if the <code>select</code> function returns a 1:many mapping then you can’t just append the annotation to the fit object. An alternative way to get annotation will be discussed tomorrow during the analysis of the second dataset.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.cont$genes &lt;-<span class="st"> </span>ann</code></pre></div>
<p>Now when we run the <code>topTable</code> command, the annotation information should be included in the output.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit.cont,<span class="dt">coef=</span><span class="st">&quot;B.PregVsLac&quot;</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>       ENTREZID        SYMBOL                                     GENENAME
24117     24117          Wif1                      Wnt inhibitory factor 1
381290   381290        Atp2b4 ATPase, Ca++ transporting, plasma membrane 4
78896     78896 1500015O10Rik                   RIKEN cDNA 1500015O10 gene
226101   226101          Myof                                    myoferlin
16012     16012        Igfbp6 insulin-like growth factor binding protein 6
231830   231830       Micall2                                 MICAL-like 2
16669     16669         Krt19                                   keratin 19
55987     55987         Cpxm2            carboxypeptidase X 2 (M14 family)
231991   231991         Creb5    cAMP responsive element binding protein 5
14620     14620          Gjb3                 gap junction protein, beta 3
           logFC  AveExpr         t      P.Value    adj.P.Val        B
24117   1.819943 2.975545  20.10780 1.063770e-10 1.016240e-06 14.96977
381290 -2.143885 3.944066 -19.07495 1.982934e-10 1.016240e-06 14.39556
78896   2.807548 3.036519  18.54773 2.758828e-10 1.016240e-06 14.07416
226101 -2.329744 6.223525 -18.26861 3.297667e-10 1.016240e-06 13.85802
16012  -2.896115 1.978449 -18.21525 3.413066e-10 1.016240e-06 13.46984
231830  2.253400 4.760597  18.02627 3.858161e-10 1.016240e-06 13.67600
16669  -2.312721 8.741892 -17.07937 7.264548e-10 1.640127e-06 13.17357
55987  -1.515469 2.834512 -16.64333 9.829870e-10 1.718703e-06 12.88782
231991 -2.598105 4.275930 -16.53634 1.059885e-09 1.718703e-06 12.66020
14620   3.600094 3.525281  16.46627 1.113755e-09 1.718703e-06 12.54071
</code></pre>
<p>To get the full table (i.e. the information for all genes, not just the top 10) we can specify <code>n=&quot;Inf&quot;</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">limma.res &lt;-<span class="st"> </span><span class="kw">topTable</span>(fit.cont,<span class="dt">coef=</span><span class="st">&quot;B.PregVsLac&quot;</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>,<span class="dt">n=</span><span class="st">&quot;Inf&quot;</span>)</code></pre></div>
<p>We can save the results table using the <code>write.csv</code> function, which writes the results out to a csv file, which you can open in excel.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.csv</span>(limma.res,<span class="dt">file=</span><span class="st">&quot;B.PregVsLacResults.csv&quot;</span>,<span class="dt">row.names=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><strong>A note about deciding how many genes are significant</strong>: In order to decide which genes are differentially expressed, we usually take a cut-off of 0.05 on the adjusted p-value, NOT the raw p-value. This is because we are testing more than 15000 genes, and the chances of finding differentially expressed genes is very high when you do that many tests. Hence we need to control the false discovery rate, which is the adjusted p-value column in the results table. What this means is that if 100 genes are significant at a 5% false discovery rate, we are willing to accept that 5 will be false positives. Note that the <code>decideTests</code> function displays significant genes at 5% FDR.</p>
<h3 id="plots-after-testing-for-de">Plots after testing for DE</h3>
<p>Let’s do a few plots to make sure everything looks good and that we haven’t made a mistake in the analysis. Genome-wide plots that are useful for checking are MAplots (or MDplots) and volcano plots. There are functions in limma for plotting these with <code>fit.cont</code> as input.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We want to highlight the significant genes. We can get this from decideTests.</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMD</span>(fit.cont,<span class="dt">coef=</span><span class="dv">1</span>,<span class="dt">status=</span>summa.fit[,<span class="st">&quot;B.PregVsLac&quot;</span>])

<span class="co"># For the volcano plot we have to specify how many of the top genes to hightlight.</span>
<span class="co"># We can also specify that we want to plot the gene symbol for the highlighted genes.</span>
<span class="co"># let&#39;s highlight the top 100 most DE genes</span>
<span class="kw">volcanoplot</span>(fit.cont,<span class="dt">coef=</span><span class="dv">1</span>,<span class="dt">highlight=</span><span class="dv">100</span>,<span class="dt">names=</span>fit.cont$genes$SYMBOL)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-55-1.png" title="plot of chunk unnamed-chunk-55" alt="plot of chunk unnamed-chunk-55" style="display: block; margin: auto;" /></p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-6"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<p>Look at the MD plot and volcano plot for the second comparison, <code>L.PregVsLac</code>. Change the number of highlighted genes to 200 in the volcano plot.</p>
</div>
</section>
<p><img src="fig-06/unnamed-chunk-56-1.png" title="plot of chunk unnamed-chunk-56" alt="plot of chunk unnamed-chunk-56" style="display: block; margin: auto;" /></p>
<p>Before following up on the DE genes with further lab work, it is recommended to have a look at the expression levels of the individual samples for the genes of interest. We can quickly look at grouped expression using <code>stripchart</code>. We can use the normalised log expression values in the voom object (<code>v$E</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))
<span class="co"># Let&#39;s look at the first gene in the topTable, Wif1, which has a rowname 24117</span>
<span class="kw">stripchart</span>(v$E[<span class="st">&quot;24117&quot;</span>,]~group)
<span class="co"># This plot is ugly, let&#39;s make it better</span>
<span class="kw">stripchart</span>(v$E[<span class="st">&quot;24117&quot;</span>,]~group,<span class="dt">vertical=</span><span class="ot">TRUE</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">cex.axis=</span><span class="fl">0.8</span>,<span class="dt">pch=</span><span class="dv">16</span>,<span class="dt">col=</span><span class="dv">1</span>:<span class="dv">6</span>,<span class="dt">method=</span><span class="st">&quot;jitter&quot;</span>)
<span class="co"># Let&#39;s use nicer colours</span>
nice.col &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">6</span>,<span class="dt">name=</span><span class="st">&quot;Dark2&quot;</span>)
<span class="kw">stripchart</span>(v$E[<span class="st">&quot;24117&quot;</span>,]~group,<span class="dt">vertical=</span><span class="ot">TRUE</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">cex.axis=</span><span class="fl">0.8</span>,<span class="dt">pch=</span><span class="dv">16</span>,<span class="dt">cex=</span><span class="fl">1.3</span>,<span class="dt">col=</span>nice.col,<span class="dt">method=</span><span class="st">&quot;jitter&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Normalised log2 expression&quot;</span>,<span class="dt">main=</span><span class="st">&quot;Wif1&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-57-1.png" title="plot of chunk unnamed-chunk-57" alt="plot of chunk unnamed-chunk-57" style="display: block; margin: auto;" /></p>
<p>Notice anything interesting about the expression of this gene?</p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-7"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<p>Take the top gene from the L.PregVsLactate comparison and make a stripchart of grouped expression as above. (Don’t forget to change the title of the plot.)</p>
</div>
</section>
<p><strong>Solution</strong></p>
<pre class="output"><code>       ENTREZID   SYMBOL
12992     12992  Csn1s2b
13358     13358  Slc25a1
20531     20531  Slc34a2
11941     11941   Atp2b2
230810   230810  Slc30a2
100705   100705    Acacb
68801     68801   Elovl5
13645     13645      Egf
12683     12683    Cidea
26366     26366 Ceacam10
                                                                              GENENAME
12992                                                           casein alpha s2-like B
13358  solute carrier family 25 (mitochondrial carrier, citrate transporter), member 1
20531                            solute carrier family 34 (sodium phosphate), member 2
11941                                     ATPase, Ca++ transporting, plasma membrane 2
230810                           solute carrier family 30 (zinc transporter), member 2
100705                                              acetyl-Coenzyme A carboxylase beta
68801              ELOVL family member 5, elongation of long chain fatty acids (yeast)
13645                                                          epidermal growth factor
12683      cell death-inducing DNA fragmentation factor, alpha subunit-like effector A
26366                       carcinoembryonic antigen-related cell adhesion molecule 10
           logFC   AveExpr         t      P.Value    adj.P.Val        B
12992  -8.603611 3.5629500 -46.84744 4.172002e-15 6.593433e-11 23.84700
13358  -4.124175 5.7796989 -34.76897 1.526326e-13 1.206103e-09 21.38712
20531  -4.177812 4.2786290 -31.24354 5.524752e-13 2.910439e-09 20.23287
11941  -7.386986 1.2821431 -30.01476 8.946195e-13 3.534642e-09 18.54201
230810 -3.203118 2.6958115 -29.31342 1.188209e-12 3.654867e-09 19.42820
100705 -4.314320 4.4409137 -28.93689 1.387573e-12 3.654867e-09 19.30790
68801  -2.863304 6.4552045 -28.18823 1.900049e-12 3.828415e-09 19.08355
13645  -5.362664 0.7359047 -28.14183 1.937947e-12 3.828415e-09 17.22920
12683  -4.840655 3.3749575 -27.04271 3.123304e-12 5.484522e-09 18.58932
26366  -3.295621 1.8210142 -25.45612 6.437086e-12 1.005946e-08 17.82256
</code></pre>
<p><img src="fig-06/unnamed-chunk-58-1.png" title="plot of chunk unnamed-chunk-58" alt="plot of chunk unnamed-chunk-58" style="display: block; margin: auto;" /></p>
<p>An interactive version of the volcano plot above that includes the raw per sample values in a separate panel is possible via the <code>glXYPlot</code> function in the <em>Glimma</em> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">group2 &lt;-<span class="st"> </span>group
<span class="kw">levels</span>(group2) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;basal.lactate&quot;</span>,<span class="st">&quot;basal.preg&quot;</span>,<span class="st">&quot;basal.virgin&quot;</span>,<span class="st">&quot;lum.lactate&quot;</span>, <span class="st">&quot;lum.preg&quot;</span>, <span class="st">&quot;lum.virgin&quot;</span>)
<span class="kw">glXYPlot</span>(<span class="dt">x=</span>fit.cont$coefficients[,<span class="dv">1</span>], <span class="dt">y=</span>fit.cont$lods[,<span class="dv">1</span>],
         <span class="dt">xlab=</span><span class="st">&quot;logFC&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;B&quot;</span>, <span class="dt">main=</span><span class="st">&quot;B.PregVsLac&quot;</span>,
         <span class="dt">counts=</span>y$counts, <span class="dt">groups=</span>group2, <span class="dt">status=</span>summa.fit[,<span class="dv">1</span>],
         <span class="dt">anno=</span>fit.cont$genes, <span class="dt">id.column=</span><span class="st">&quot;ENTREZID&quot;</span>, <span class="dt">folder=</span><span class="st">&quot;volcano&quot;</span>)</code></pre></div>
<p>This function creates an html page (./volcano/XY-Plot.html) with a volcano plot on the left and a plot showing the log-CPM per sample for a selected gene on the right. A search bar is available to search for genes of interest.</p>
<h3 id="testing-relative-to-a-threshold-treat">Testing relative to a threshold (TREAT)</h3>
<p>When there is a lot of differential expression, sometimes we may want to cut-off on a fold change threshold as well as a p-value threshold so that we follow up on the most biologically significant genes. However, it is not recommended to simply rank by p-value and then discard genes with small logFC’s, as this has been shown to increase the false discovery rate. In other words, you are not controlling the false discovery rate at 5% anymore. There is a function called <code>treat</code> in the <em>limma</em> package that performs this style of analysis correctly <span class="citation">(D. J. McCarthy and Smyth 2009)</span>. <code>treat</code> will simply take our <code>fit.cont</code> object, as well as a user-specified log fold change cut-off, and recalculate the moderated t-statistics and p-values with the new information about logFC.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Let&#39;s decide that we are only interested in genes that have a absolute logFC of 1.</span>
<span class="co"># This corresponds to a fold change of 2, or 0.5 (i.e. double or half).</span>
<span class="co"># We can perform a treat analysis which ranks our genes according to p-value AND logFC.</span>
<span class="co"># This is easy to do after our analysis, we just give the treat function the fit.cont object and specify our cut-off.</span>
fit.treat &lt;-<span class="st"> </span><span class="kw">treat</span>(fit.cont,<span class="dt">lfc=</span><span class="dv">1</span>)
res.treat &lt;-<span class="st"> </span><span class="kw">decideTests</span>(fit.treat)
<span class="kw">summary</span>(res.treat)</code></pre></div>
<pre class="output"><code>   B.PregVsLac L.PregVsLac
-1          53         631
0        15677       14775
1           74         398
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit.treat,<span class="dt">coef=</span><span class="dv">1</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>       ENTREZID        SYMBOL                                     GENENAME
211577   211577        Mrgprf                    MAS-related GPR, member F
78896     78896 1500015O10Rik                   RIKEN cDNA 1500015O10 gene
16012     16012        Igfbp6 insulin-like growth factor binding protein 6
21953     21953         Tnni2                 troponin I, skeletal, fast 2
14620     14620          Gjb3                 gap junction protein, beta 3
12992     12992       Csn1s2b                       casein alpha s2-like B
270150   270150       Ccdc153            coiled-coil domain containing 153
226101   226101          Myof                                    myoferlin
381290   381290        Atp2b4 ATPase, Ca++ transporting, plasma membrane 4
231991   231991         Creb5    cAMP responsive element binding protein 5
           logFC    AveExpr         t      P.Value    adj.P.Val
211577 -5.146268 -0.9368335 -13.18562 7.329299e-09 6.669358e-05
78896   2.807548  3.0365195  11.94135 2.234154e-08 6.669358e-05
16012  -2.896115  1.9784488 -11.92570 2.267536e-08 6.669358e-05
21953  -5.827889  0.3020716 -11.93183 2.295775e-08 6.669358e-05
14620   3.600094  3.5252805  11.89242 2.342636e-08 6.669358e-05
12992  -6.070143  3.5629500 -11.83199 2.532027e-08 6.669358e-05
270150 -3.211148 -1.3408388 -10.67393 7.828123e-08 1.767367e-04
226101 -2.329744  6.2235246 -10.42714 1.012983e-07 2.001148e-04
381290 -2.143885  3.9440659 -10.17758 1.322118e-07 2.050763e-04
231991 -2.598105  4.2759295 -10.17157 1.330809e-07 2.050763e-04
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Notice that much fewer genes are highlighted in the MAplot</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMD</span>(fit.treat,<span class="dt">coef=</span><span class="dv">1</span>,<span class="dt">status=</span>res.treat[,<span class="st">&quot;B.PregVsLac&quot;</span>])
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)
<span class="kw">plotMD</span>(fit.treat,<span class="dt">coef=</span><span class="dv">2</span>,<span class="dt">status=</span>res.treat[,<span class="st">&quot;L.PregVsLac&quot;</span>])
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-60-1.png" title="plot of chunk unnamed-chunk-60" alt="plot of chunk unnamed-chunk-60" style="display: block; margin: auto;" /></p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-8"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<p>Change the cut-off so that we are interested in genes that change at least 50% on the fold change scale.</p>
<p>HINT: what is the corresponding logFC value of 50% fold change? Assume basal.pregnant is 50% higher than basal.lactate</p>
</div>
</section>
<p><strong>Solution</strong></p>
<pre class="output"><code>   B.PregVsLac L.PregVsLac
-1         234        1328
0        15250       13397
1          320        1079
</code></pre>
<pre class="output"><code>       ENTREZID        SYMBOL                                     GENENAME
78896     78896 1500015O10Rik                   RIKEN cDNA 1500015O10 gene
16012     16012        Igfbp6 insulin-like growth factor binding protein 6
211577   211577        Mrgprf                    MAS-related GPR, member F
381290   381290        Atp2b4 ATPase, Ca++ transporting, plasma membrane 4
14620     14620          Gjb3                 gap junction protein, beta 3
226101   226101          Myof                                    myoferlin
24117     24117          Wif1                      Wnt inhibitory factor 1
231830   231830       Micall2                                 MICAL-like 2
21953     21953         Tnni2                 troponin I, skeletal, fast 2
231991   231991         Creb5    cAMP responsive element binding protein 5
           logFC    AveExpr         t      P.Value    adj.P.Val
78896   2.807548  3.0365195  14.68324 2.122741e-09 1.107588e-05
16012  -2.896115  1.9784488 -14.53610 2.387463e-09 1.107588e-05
211577 -5.146268 -0.9368335 -14.50548 2.595874e-09 1.107588e-05
381290 -2.143885  3.9440659 -13.87032 4.068604e-09 1.107588e-05
14620   3.600094  3.5252805  13.79074 4.435571e-09 1.107588e-05
226101 -2.329744  6.2235246 -13.68164 4.766155e-09 1.107588e-05
24117   1.819943  2.9755445  13.64479 4.905791e-09 1.107588e-05
231830  2.253400  4.7605970  13.34681 6.327022e-09 1.249903e-05
21953  -5.827889  0.3020716 -12.95757 9.721628e-09 1.399330e-05
231991 -2.598105  4.2759295 -12.81319 1.010560e-08 1.399330e-05
</code></pre>
<p>An interactive version of the mean-difference plots is possible via the <code>glMDPlot</code> function in the <em>Glimma</em> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">glMDPlot</span>(fit.treat, <span class="dt">coef=</span><span class="dv">1</span>, <span class="dt">counts=</span>y$counts, <span class="dt">groups=</span>group2,
        <span class="dt">status=</span>res.treat, <span class="dt">id.column=</span><span class="st">&quot;ENTREZID&quot;</span>, <span class="dt">main=</span><span class="st">&quot;B.PregVsLac&quot;</span>,
        <span class="dt">folder=</span><span class="st">&quot;md&quot;</span>)</code></pre></div>
<p>As with the volcano plot example above, this function creates an html page (./md/MD-Plot.html) that allows the user to search for their favourite gene.</p>
<h2 id="gene-set-testing">Gene Set Testing</h2>
<p>Sometimes there is quite a long list of differentially expressed genes to interpret after a differential expression analysis, and it is usually infeasible to go through the list one gene at a time trying to understand it’s biological function. A common downstream procedure is gene set testing, which aims to understand which pathways/gene networks the differentially expressed genes are implicated in.</p>
<p>There are a number of different ways to go about testing for enrichment of biological pathways, and the test you choose usually depends on the question you’re asking. There are two kinds of tests: self-contained and competitive gene set tests. Self-contained tests, which include the <code>ROAST</code> procedure, ask the question “Are the genes in the set/pathway differentially expressed as a whole?” Competitive gene set tests, like <code>goana</code> and <code>camera</code> ask the question whether the differentially expressed genes tend to be over-represented in the gene set, compared to all the other genes in the experiment. These different questions use different statistical methodology.</p>
<h3 id="gene-ontology-testing-with-goana">Gene ontology testing with goana</h3>
<p>First, we will perform a gene ontology (GO) enrichment analysis using the <code>goana</code> function in <em>limma</em>. There are approximately 20,000 GO terms, and they are split into three categories: BP (biological process), MF (molecular function) and CC (cellular component). <code>goana</code> uses annotation from the appropriate Bioconductor package and can be used for any of the five species specified (Hs, Mm, Rn, Dm or Pt). <code>goana</code> has an advantage over other methods, such as DAVID, in that there is the option to take into account the gene length bias inherent in RNA-Seq data.</p>
<p>Suppose we want to identify GO terms that are over-represented in the basal lactating group compared to the basal pregnancy group. This can be achieved by applying the goana function to the differential expression results of that comparison. <code>goana</code> takes the <code>fit.cont</code> object, the coefficient of interest and the species. The top set of most enriched GO terms can be viewed with the topGO function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">go &lt;-<span class="st"> </span><span class="kw">goana</span>(fit.cont, <span class="dt">coef=</span><span class="st">&quot;B.PregVsLac&quot;</span>,<span class="dt">species =</span> <span class="st">&quot;Mm&quot;</span>)
<span class="kw">topGO</span>(go, <span class="dt">n=</span><span class="dv">10</span>)</code></pre></div>
<pre class="output"><code>                                              Term Ont    N  Up Down
GO:0044822                     poly(A) RNA binding  MF 1098 406  114
GO:0003723                             RNA binding  MF 1461 482  162
GO:0030529 intracellular ribonucleoprotein complex  CC  686 273   53
GO:1990904               ribonucleoprotein complex  CC  686 273   53
GO:0022613    ribonucleoprotein complex biogenesis  BP  361 176   24
GO:0042254                     ribosome biogenesis  BP  247 138    8
GO:0022626                      cytosolic ribosome  CC  102  79    1
GO:0003676                    nucleic acid binding  MF 2878 719  438
GO:0006364                         rRNA processing  BP  168  97    3
GO:0016072                  rRNA metabolic process  BP  171  97    5
                   P.Up    P.Down
GO:0044822 2.877931e-61 1.0000000
GO:0003723 1.127696e-55 1.0000000
GO:0030529 1.493472e-47 1.0000000
GO:1990904 1.493472e-47 1.0000000
GO:0022613 7.457785e-45 1.0000000
GO:0042254 4.873452e-44 1.0000000
GO:0022626 5.987149e-41 1.0000000
GO:0003676 1.052725e-33 0.9999215
GO:0006364 7.593316e-33 1.0000000
GO:0016072 5.598207e-32 1.0000000
</code></pre>
<p>The row names of the output are the universal identifiers of the GO terms, with one term per row. The Term column gives the names of the GO terms. These terms cover three domains - biological process (BP), cellular component (CC) and molecular function (MF), as shown in the Ont column. The N column represents the total number of genes that are annotated with each GO term. The Up and Down columns represent the number of differentially expressed genes that overlap with the genes in the GO term. The P.Up and P.Down columns contain the p-values for over-representation of the GO term across the set of up- and down-regulated genes, respectively. The output table is sorted by the minimum of P.Up and P.Down by default.</p>
<p>An additional refinement is to supply <code>goana</code> with the gene lengths using the <code>covariate</code> argument. In the original data matrix that we loaded into R, there is a column called “Length”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(seqdata)</code></pre></div>
<pre class="output"><code> [1] &quot;EntrezGeneID&quot;                     
 [2] &quot;Length&quot;                           
 [3] &quot;MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1&quot;
 [4] &quot;MCL1.DH_BC2CTUACXX_CAGATC_L002_R1&quot;
 [5] &quot;MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1&quot;
 [6] &quot;MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1&quot;
 [7] &quot;MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1&quot;
 [8] &quot;MCL1.DL_BC2CTUACXX_ATCACG_L002_R1&quot;
 [9] &quot;MCL1.LA_BC2CTUACXX_GATCAG_L001_R1&quot;
[10] &quot;MCL1.LB_BC2CTUACXX_TGACCA_L001_R1&quot;
[11] &quot;MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1&quot;
[12] &quot;MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1&quot;
[13] &quot;MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1&quot;
[14] &quot;MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1&quot;
</code></pre>
<p>In order to get the gene lengths for every gene in <code>fit.cont</code>, we can use the <code>match</code> command. Note that the gene length supplied needs to be in the correct order.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">rownames</span>(fit.cont),seqdata$EntrezGeneID)
gene_length &lt;-<span class="st"> </span>seqdata$Length[m]
<span class="kw">head</span>(gene_length)</code></pre></div>
<pre class="output"><code>[1] 3634 3130 4203 2433 2847 2241
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Rerun goana with gene length information</span>
go_length &lt;-<span class="st"> </span><span class="kw">goana</span>(fit.cont,<span class="dt">coef=</span><span class="st">&quot;B.PregVsLac&quot;</span>,<span class="dt">species=</span><span class="st">&quot;Mm&quot;</span>,<span class="dt">covariate=</span>gene_length)
<span class="kw">topGO</span>(go_length, <span class="dt">n=</span><span class="dv">10</span>)</code></pre></div>
<pre class="output"><code>                                              Term Ont    N  Up Down
GO:0044822                     poly(A) RNA binding  MF 1098 406  114
GO:0003723                             RNA binding  MF 1461 482  162
GO:0030529 intracellular ribonucleoprotein complex  CC  686 273   53
GO:1990904               ribonucleoprotein complex  CC  686 273   53
GO:0042254                     ribosome biogenesis  BP  247 138    8
GO:0022613    ribonucleoprotein complex biogenesis  BP  361 176   24
GO:0022626                      cytosolic ribosome  CC  102  79    1
GO:0006364                         rRNA processing  BP  168  97    3
GO:0016072                  rRNA metabolic process  BP  171  97    5
GO:0005840                                ribosome  CC  205 104    5
                   P.Up P.Down
GO:0044822 6.660226e-62      1
GO:0003723 7.742138e-57      1
GO:0030529 2.708474e-51      1
GO:1990904 2.708474e-51      1
GO:0042254 3.722242e-47      1
GO:0022613 6.822355e-47      1
GO:0022626 8.920719e-46      1
GO:0006364 6.146939e-35      1
GO:0016072 6.543744e-34      1
GO:0005840 1.381249e-33      1
</code></pre>
<p>Notes</p>
<ul>
<li>Users can specify the domain of the enriched GO terms in topGO. For instance, topGO(go,ontology=“BP”) lists the top set of most enriched GO terms that are related to a biological process. This avoids other domains that are not of interest.</li>
<li>The goana function uses the NCBI RefSeq annotation. Therefore, the Entrez Gene identifier (ID) should be supplied for each gene as the row names of the fit object.</li>
<li>Users should set species according to the organism being studied.</li>
</ul>
<h3 id="camera-gene-set-testing-using-the-broads-curated-gene-sets">CAMERA gene set testing using the Broad’s curated gene sets</h3>
<p>Other databases of gene sets that are available come from the Broad Institute’s Molecular Signatures Database (<a href="http://software.broadinstitute.org/gsea/msigdb">MSigDB</a>). CAMERA is good option for testing a very large number of gene sets such as the MSigDB sets, as it is very fast. CAMERA is known as a competitive gene set test, however it has the advantage that it can take into account inter-gene correlation within each gene set <span class="citation">(D. Wu and Smyth 2012)</span>. It also works seemlessly with a <code>voom</code> object, taking into account the mean-variance relationship in RNA-Seq data.</p>
<p>Here we will be using the C2 gene sets for mouse, available as .rdata files from the WEHI bioinformatics page <a href="http://bioinf.wehi.edu.au/software/MSigDB/index.html" class="uri">http://bioinf.wehi.edu.au/software/MSigDB/index.html</a>. The C2 gene sets contain 4725 curated gene sets collected from a variety of places: BioCarta, KEGG, Pathway Interaction Database, Reactome as well as some published studies.It doesn’t include GO terms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load in the mouse c2 gene sets</span>
<span class="co"># The R object is called Mm.c2</span>
<span class="kw">load</span>(<span class="st">&quot;data/mouse_c2_v5.rdata&quot;</span>)
<span class="co"># Have a look at the first few gene sets</span>
<span class="kw">names</span>(Mm.c2)[<span class="dv">1</span>:<span class="dv">5</span>]</code></pre></div>
<pre class="output"><code>[1] &quot;KEGG_GLYCOLYSIS_GLUCONEOGENESIS&quot;              
[2] &quot;KEGG_CITRATE_CYCLE_TCA_CYCLE&quot;                 
[3] &quot;KEGG_PENTOSE_PHOSPHATE_PATHWAY&quot;               
[4] &quot;KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS&quot;
[5] &quot;KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM&quot;         
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Number of gene sets in C2</span>
<span class="kw">length</span>(Mm.c2)</code></pre></div>
<pre class="output"><code>[1] 4725
</code></pre>
<p>The gene identifiers are Entrez Gene ID, which is the same as the rownames of our voom object. We need to map the Entrez gene ids between the list of gene sets and our voom object. We can do this using the <code>ids2indices</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">c2.ind &lt;-<span class="st"> </span><span class="kw">ids2indices</span>(Mm.c2, <span class="kw">rownames</span>(v))</code></pre></div>
<p>CAMERA takes as input the voom object <code>v</code>, the indexed list of gene sets <code>c2.ind</code>, the design matrix, the contrast being tested, as well as some other arguments. By default, CAMERA can estimate the correlation for each gene set separately. However, in practise, it works well to set a small inter-gene correlation of about 0.05 using the <code>inter.gene.cor</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gst.camera &lt;-<span class="st"> </span><span class="kw">camera</span>(v,<span class="dt">index=</span>c2.ind,<span class="dt">design=</span>design,<span class="dt">contrast =</span> cont.matrix[,<span class="dv">1</span>],<span class="dt">inter.gene.cor=</span><span class="fl">0.05</span>)</code></pre></div>
<p>CAMERA outputs a dataframe of the resulting statistics, with each row denoting a different gene set. The output is ordered by p-value so that the most significant should be at the top. Let’s look at the top 5 gene sets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gst.camera[<span class="dv">1</span>:<span class="dv">5</span>,]</code></pre></div>
<pre class="output"><code>                                                                           NGenes
REACTOME_PEPTIDE_CHAIN_ELONGATION                                              81
KEGG_RIBOSOME                                                                  83
REACTOME_3_UTR_MEDIATED_TRANSLATIONAL_REGULATION                              102
REACTOME_FORMATION_OF_THE_TERNARY_COMPLEX_AND_SUBSEQUENTLY_THE_43S_COMPLEX     48
REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE          108
                                                                           Direction
REACTOME_PEPTIDE_CHAIN_ELONGATION                                                 Up
KEGG_RIBOSOME                                                                     Up
REACTOME_3_UTR_MEDIATED_TRANSLATIONAL_REGULATION                                  Up
REACTOME_FORMATION_OF_THE_TERNARY_COMPLEX_AND_SUBSEQUENTLY_THE_43S_COMPLEX        Up
REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE              Up
                                                                                 PValue
REACTOME_PEPTIDE_CHAIN_ELONGATION                                          3.711614e-10
KEGG_RIBOSOME                                                              4.242401e-10
REACTOME_3_UTR_MEDIATED_TRANSLATIONAL_REGULATION                           1.484636e-09
REACTOME_FORMATION_OF_THE_TERNARY_COMPLEX_AND_SUBSEQUENTLY_THE_43S_COMPLEX 1.477539e-08
REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE       1.716413e-08
                                                                                    FDR
REACTOME_PEPTIDE_CHAIN_ELONGATION                                          1.001631e-06
KEGG_RIBOSOME                                                              1.001631e-06
REACTOME_3_UTR_MEDIATED_TRANSLATIONAL_REGULATION                           2.336816e-06
REACTOME_FORMATION_OF_THE_TERNARY_COMPLEX_AND_SUBSEQUENTLY_THE_43S_COMPLEX 1.620981e-05
REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE       1.620981e-05
</code></pre>
<p>The total number of significant gene sets at 5% FDR is</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(gst.camera$FDR &lt;<span class="st"> </span><span class="fl">0.05</span>)</code></pre></div>
<pre class="output"><code>
FALSE  TRUE 
 4593   129 
</code></pre>
<p>You can write out the camera results to a csv file to open in excel.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.csv</span>(gst.camera,<span class="dt">file=</span><span class="st">&quot;gst_BPregVsLac.csv&quot;</span>)</code></pre></div>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-9"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Run <code>camera</code> on the second contrast in the contrast matrix.</li>
<li>Run <code>camera</code> on a different set of MSigDB gene sets, the hallmark datasets, <code>mouse_H_v5.rdata</code>. You will need to load in the hallmark gene sets, and the object will be called <code>Mm.H</code> in R.</li>
</ol>
</div>
</section>
<p><strong>Solution</strong></p>
<pre class="output"><code>
FALSE  TRUE 
   45     5 
</code></pre>
<pre class="output"><code>                                   NGenes Direction       PValue
HALLMARK_MYC_TARGETS_V2                60        Up 7.859018e-09
HALLMARK_MYC_TARGETS_V1               204        Up 3.706364e-07
HALLMARK_E2F_TARGETS                  201        Up 1.198872e-04
HALLMARK_G2M_CHECKPOINT               199        Up 6.547050e-04
HALLMARK_OXIDATIVE_PHOSPHORYLATION    199        Up 2.493738e-03
HALLMARK_TGF_BETA_SIGNALING            53        Up 1.068929e-02
HALLMARK_UNFOLDED_PROTEIN_RESPONSE    110        Up 1.669518e-02
HALLMARK_CHOLESTEROL_HOMEOSTASIS       73      Down 1.956095e-02
HALLMARK_ALLOGRAFT_REJECTION          159        Up 5.451666e-02
HALLMARK_MTORC1_SIGNALING             201        Up 1.088344e-01
                                            FDR
HALLMARK_MYC_TARGETS_V2            3.929509e-07
HALLMARK_MYC_TARGETS_V1            9.265909e-06
HALLMARK_E2F_TARGETS               1.998120e-03
HALLMARK_G2M_CHECKPOINT            8.183812e-03
HALLMARK_OXIDATIVE_PHOSPHORYLATION 2.493738e-02
HALLMARK_TGF_BETA_SIGNALING        8.907745e-02
HALLMARK_UNFOLDED_PROTEIN_RESPONSE 1.192513e-01
HALLMARK_CHOLESTEROL_HOMEOSTASIS   1.222560e-01
HALLMARK_ALLOGRAFT_REJECTION       3.028704e-01
HALLMARK_MTORC1_SIGNALING          5.441718e-01
</code></pre>
<h3 id="roast-gene-set-testing">ROAST gene set testing</h3>
<p>ROAST is an example of a self-contained gene set test <span class="citation">(D. Wu et al. 2010)</span>. It asks the question, “Do the genes in my set tend to be differentially expressed between my conditions of interest?”. ROAST doesn’t care about what the other genes in the experiment are doing, which is different to <code>camera</code> and <code>goana</code>. ROAST is a good option for when you’re interested in a specific set, or a few sets. It is not really used to test thousands of sets at one time.</p>
<p>From the Hallmark gene sets, two MYC pathways were most significant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">H.camera[<span class="dv">1</span>:<span class="dv">10</span>,]</code></pre></div>
<pre class="output"><code>                                   NGenes Direction       PValue
HALLMARK_MYC_TARGETS_V2                60        Up 7.859018e-09
HALLMARK_MYC_TARGETS_V1               204        Up 3.706364e-07
HALLMARK_E2F_TARGETS                  201        Up 1.198872e-04
HALLMARK_G2M_CHECKPOINT               199        Up 6.547050e-04
HALLMARK_OXIDATIVE_PHOSPHORYLATION    199        Up 2.493738e-03
HALLMARK_TGF_BETA_SIGNALING            53        Up 1.068929e-02
HALLMARK_UNFOLDED_PROTEIN_RESPONSE    110        Up 1.669518e-02
HALLMARK_CHOLESTEROL_HOMEOSTASIS       73      Down 1.956095e-02
HALLMARK_ALLOGRAFT_REJECTION          159        Up 5.451666e-02
HALLMARK_MTORC1_SIGNALING             201        Up 1.088344e-01
                                            FDR
HALLMARK_MYC_TARGETS_V2            3.929509e-07
HALLMARK_MYC_TARGETS_V1            9.265909e-06
HALLMARK_E2F_TARGETS               1.998120e-03
HALLMARK_G2M_CHECKPOINT            8.183812e-03
HALLMARK_OXIDATIVE_PHOSPHORYLATION 2.493738e-02
HALLMARK_TGF_BETA_SIGNALING        8.907745e-02
HALLMARK_UNFOLDED_PROTEIN_RESPONSE 1.192513e-01
HALLMARK_CHOLESTEROL_HOMEOSTASIS   1.222560e-01
HALLMARK_ALLOGRAFT_REJECTION       3.028704e-01
HALLMARK_MTORC1_SIGNALING          5.441718e-01
</code></pre>
<p>Let’s see if there are any MYC signalling pathways in MsigDB C2 collection. We can do this with the <code>grep</code> command on the names of the gene sets.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">grep</span>(<span class="st">&quot;MYC_&quot;</span>,<span class="kw">names</span>(c2.ind))</code></pre></div>
<pre class="output"><code> [1]  496  544  621 1591 1592 1783 1784 2029 2030 2031 2032 2194 2195 2203
[15] 2204 2435 2436 2506 2507 2600 2609 2620 2621 2629 2676 2694 2706 2716
[29] 2727 2744 2786 2824 3311 3401 3817 3943 3944 4204 4205 4206 4207 4433
[43] 4434 4616
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Let&#39;s save these so that we can subset c2.ind to test all gene sets with MYC in the name</span>
myc &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;MYC_&quot;</span>,<span class="kw">names</span>(c2.ind))
<span class="co"># What are these pathways called?</span>
<span class="kw">names</span>(c2.ind)[myc]</code></pre></div>
<pre class="output"><code> [1] &quot;PID_MYC_ACTIV_PATHWAY&quot;                          
 [2] &quot;PID_MYC_PATHWAY&quot;                                
 [3] &quot;PID_MYC_REPRESS_PATHWAY&quot;                        
 [4] &quot;ODONNELL_TARGETS_OF_MYC_AND_TFRC_UP&quot;            
 [5] &quot;ODONNELL_TARGETS_OF_MYC_AND_TFRC_DN&quot;            
 [6] &quot;CAIRO_PML_TARGETS_BOUND_BY_MYC_UP&quot;              
 [7] &quot;CAIRO_PML_TARGETS_BOUND_BY_MYC_DN&quot;              
 [8] &quot;SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY&quot;       
 [9] &quot;SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM&quot;       
[10] &quot;SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_DN&quot;    
[11] &quot;SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP&quot;    
[12] &quot;CEBALLOS_TARGETS_OF_TP53_AND_MYC_UP&quot;            
[13] &quot;CEBALLOS_TARGETS_OF_TP53_AND_MYC_DN&quot;            
[14] &quot;KIM_MYC_AMPLIFICATION_TARGETS_UP&quot;               
[15] &quot;KIM_MYC_AMPLIFICATION_TARGETS_DN&quot;               
[16] &quot;BENPORATH_MYC_TARGETS_WITH_EBOX&quot;                
[17] &quot;BENPORATH_MYC_MAX_TARGETS&quot;                      
[18] &quot;MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_UP&quot;         
[19] &quot;MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN&quot;         
[20] &quot;SCHUHMACHER_MYC_TARGETS_DN&quot;                     
[21] &quot;LEE_LIVER_CANCER_MYC_TGFA_DN&quot;                   
[22] &quot;LEE_LIVER_CANCER_MYC_UP&quot;                        
[23] &quot;MENSSEN_MYC_TARGETS&quot;                            
[24] &quot;LEE_LIVER_CANCER_MYC_TGFA_UP&quot;                   
[25] &quot;LEE_LIVER_CANCER_MYC_E2F1_DN&quot;                   
[26] &quot;LEE_LIVER_CANCER_MYC_E2F1_UP&quot;                   
[27] &quot;SCHUHMACHER_MYC_TARGETS_UP&quot;                     
[28] &quot;LEE_LIVER_CANCER_MYC_DN&quot;                        
[29] &quot;COLLER_MYC_TARGETS_UP&quot;                          
[30] &quot;COLLER_MYC_TARGETS_DN&quot;                          
[31] &quot;YU_MYC_TARGETS_UP&quot;                              
[32] &quot;YU_MYC_TARGETS_DN&quot;                              
[33] &quot;BILD_MYC_ONCOGENIC_SIGNATURE&quot;                   
[34] &quot;SANSOM_APC_MYC_TARGETS&quot;                         
[35] &quot;SOUCEK_MYC_TARGETS&quot;                             
[36] &quot;ELLWOOD_MYC_TARGETS_UP&quot;                         
[37] &quot;ELLWOOD_MYC_TARGETS_DN&quot;                         
[38] &quot;DANG_REGULATED_BY_MYC_UP&quot;                       
[39] &quot;DANG_REGULATED_BY_MYC_DN&quot;                       
[40] &quot;DANG_MYC_TARGETS_UP&quot;                            
[41] &quot;DANG_MYC_TARGETS_DN&quot;                            
[42] &quot;ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_UP&quot;
[43] &quot;ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_DN&quot;
[44] &quot;ALFANO_MYC_TARGETS&quot;                             
</code></pre>
<p>Let’s use ROAST to see if these MYC related gene sets tend to be differentially expressed. Note that the syntax for <code>camera</code> and <code>roast</code> is almost identical.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myc.rst &lt;-<span class="st"> </span><span class="kw">roast</span>(v,<span class="dt">index=</span>c2.ind[myc],<span class="dt">design=</span>design,<span class="dt">contrast=</span>cont.matrix[,<span class="dv">1</span>],<span class="dt">nrot=</span><span class="dv">999</span>)
myc.rst[<span class="dv">1</span>:<span class="dv">15</span>,]</code></pre></div>
<pre class="output"><code>                                                NGenes   PropDown
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_DN         48 0.06250000
MENSSEN_MYC_TARGETS                                 54 0.11111111
CAIRO_PML_TARGETS_BOUND_BY_MYC_UP                   24 0.08333333
DANG_MYC_TARGETS_UP                                144 0.09722222
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP         48 0.12500000
PID_MYC_ACTIV_PATHWAY                               79 0.12658228
SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM           156 0.12820513
LEE_LIVER_CANCER_MYC_UP                             50 0.30000000
SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY            32 0.15625000
KIM_MYC_AMPLIFICATION_TARGETS_UP                   176 0.14772727
BENPORATH_MYC_TARGETS_WITH_EBOX                    216 0.21759259
ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_DN     96 0.40625000
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN              16 0.37500000
COLLER_MYC_TARGETS_UP                               24 0.04166667
DANG_REGULATED_BY_MYC_UP                            69 0.10144928
                                                   PropUp Direction PValue
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_DN     0.7916667        Up  0.001
MENSSEN_MYC_TARGETS                             0.7592593        Up  0.001
CAIRO_PML_TARGETS_BOUND_BY_MYC_UP               0.7500000        Up  0.001
DANG_MYC_TARGETS_UP                             0.7013889        Up  0.001
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP     0.6458333        Up  0.001
PID_MYC_ACTIV_PATHWAY                           0.6455696        Up  0.001
SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM        0.6217949        Up  0.001
LEE_LIVER_CANCER_MYC_UP                         0.5400000        Up  0.001
SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY        0.5312500        Up  0.001
KIM_MYC_AMPLIFICATION_TARGETS_UP                0.5284091        Up  0.001
BENPORATH_MYC_TARGETS_WITH_EBOX                 0.4259259        Up  0.001
ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_DN 0.1875000      Down  0.001
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN          0.3125000      Down  0.001
COLLER_MYC_TARGETS_UP                           0.8333333        Up  0.002
DANG_REGULATED_BY_MYC_UP                        0.7101449        Up  0.002
                                                        FDR PValue.Mixed
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_DN     0.001692308        0.001
MENSSEN_MYC_TARGETS                             0.001692308        0.003
CAIRO_PML_TARGETS_BOUND_BY_MYC_UP               0.001692308        0.001
DANG_MYC_TARGETS_UP                             0.001692308        0.001
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP     0.001692308        0.001
PID_MYC_ACTIV_PATHWAY                           0.001692308        0.001
SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM        0.001692308        0.001
LEE_LIVER_CANCER_MYC_UP                         0.001692308        0.001
SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY        0.001692308        0.001
KIM_MYC_AMPLIFICATION_TARGETS_UP                0.001692308        0.001
BENPORATH_MYC_TARGETS_WITH_EBOX                 0.001692308        0.001
ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_DN 0.001692308        0.001
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN          0.001692308        0.001
COLLER_MYC_TARGETS_UP                           0.003882353        0.001
DANG_REGULATED_BY_MYC_UP                        0.003882353        0.002
                                                FDR.Mixed
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_DN         0.001
MENSSEN_MYC_TARGETS                                 0.003
CAIRO_PML_TARGETS_BOUND_BY_MYC_UP                   0.001
DANG_MYC_TARGETS_UP                                 0.001
SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP         0.001
PID_MYC_ACTIV_PATHWAY                               0.001
SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM            0.001
LEE_LIVER_CANCER_MYC_UP                             0.001
SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY            0.001
KIM_MYC_AMPLIFICATION_TARGETS_UP                    0.001
BENPORATH_MYC_TARGETS_WITH_EBOX                     0.001
ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_DN     0.001
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN              0.001
COLLER_MYC_TARGETS_UP                               0.001
DANG_REGULATED_BY_MYC_UP                            0.002
</code></pre>
<p>Each row corresponds to a single gene set. The NGenes column gives the number of genes in each set. The PropDown and PropUp columns contain the proportions of genes in the set that are down- and up-regulated, respectively, with absolute fold changes greater than 2. The net direction of change is determined from the significance of changes in each direction, and is shown in the Direction column. The PValue provides evidence for whether the majority of genes in the set are DE in the specified direction, whereas the PValue.Mixed tests for differential expression in any direction. FDRs are computed from the corresponding p-values across all sets.</p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-10"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Test whether the MYC signalling pathways tend to be differentially expressed between luminal pregnant vs lactating (the second contrast).</li>
<li>Look for gene sets containing “WNT” in the name and see whether they tend to be differentially expressed in basal pregnant vs lactating.</li>
</ol>
</div>
</section>
<p><strong>Solution</strong></p>
<pre class="output"><code>                                       NGenes  PropDown     PropUp
ELLWOOD_MYC_TARGETS_DN                     40 0.5750000 0.25000000
DANG_MYC_TARGETS_DN                        30 0.1666667 0.56666667
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN     16 0.5625000 0.31250000
PID_MYC_PATHWAY                            25 0.2400000 0.48000000
LEE_LIVER_CANCER_MYC_E2F1_DN               42 0.4761905 0.23809524
PID_MYC_REPRESS_PATHWAY                    59 0.2542373 0.45762712
KIM_MYC_AMPLIFICATION_TARGETS_DN           82 0.3048780 0.45121951
LEE_LIVER_CANCER_MYC_TGFA_DN               44 0.4318182 0.27272727
CEBALLOS_TARGETS_OF_TP53_AND_MYC_UP        22 0.3636364 0.09090909
CEBALLOS_TARGETS_OF_TP53_AND_MYC_DN        33 0.3030303 0.45454545
BENPORATH_MYC_TARGETS_WITH_EBOX           216 0.3194444 0.43518519
ALFANO_MYC_TARGETS                        220 0.3090909 0.42272727
ODONNELL_TARGETS_OF_MYC_AND_TFRC_DN        45 0.6666667 0.11111111
BILD_MYC_ONCOGENIC_SIGNATURE              187 0.2994652 0.45989305
YU_MYC_TARGETS_UP                          43 0.7209302 0.11627907
                                       Direction PValue         FDR
ELLWOOD_MYC_TARGETS_DN                      Down  0.001 0.002444444
DANG_MYC_TARGETS_DN                           Up  0.001 0.002444444
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN      Down  0.001 0.002444444
PID_MYC_PATHWAY                               Up  0.001 0.002444444
LEE_LIVER_CANCER_MYC_E2F1_DN                Down  0.001 0.002444444
PID_MYC_REPRESS_PATHWAY                       Up  0.001 0.002444444
KIM_MYC_AMPLIFICATION_TARGETS_DN              Up  0.001 0.002444444
LEE_LIVER_CANCER_MYC_TGFA_DN                Down  0.001 0.002444444
CEBALLOS_TARGETS_OF_TP53_AND_MYC_UP         Down  0.001 0.002444444
CEBALLOS_TARGETS_OF_TP53_AND_MYC_DN           Up  0.002 0.005500000
BENPORATH_MYC_TARGETS_WITH_EBOX               Up  0.002 0.005500000
ALFANO_MYC_TARGETS                            Up  0.002 0.005500000
ODONNELL_TARGETS_OF_MYC_AND_TFRC_DN         Down  0.005 0.015230769
BILD_MYC_ONCOGENIC_SIGNATURE                  Up  0.006 0.017285714
YU_MYC_TARGETS_UP                           Down  0.009 0.024933333
                                       PValue.Mixed FDR.Mixed
ELLWOOD_MYC_TARGETS_DN                        0.001     0.001
DANG_MYC_TARGETS_DN                           0.001     0.001
MORI_EMU_MYC_LYMPHOMA_BY_ONSET_TIME_DN        0.001     0.001
PID_MYC_PATHWAY                               0.001     0.001
LEE_LIVER_CANCER_MYC_E2F1_DN                  0.001     0.001
PID_MYC_REPRESS_PATHWAY                       0.001     0.001
KIM_MYC_AMPLIFICATION_TARGETS_DN              0.001     0.001
LEE_LIVER_CANCER_MYC_TGFA_DN                  0.001     0.001
CEBALLOS_TARGETS_OF_TP53_AND_MYC_UP           0.001     0.001
CEBALLOS_TARGETS_OF_TP53_AND_MYC_DN           0.001     0.001
BENPORATH_MYC_TARGETS_WITH_EBOX               0.001     0.001
ALFANO_MYC_TARGETS                            0.001     0.001
ODONNELL_TARGETS_OF_MYC_AND_TFRC_DN           0.002     0.002
BILD_MYC_ONCOGENIC_SIGNATURE                  0.001     0.001
YU_MYC_TARGETS_UP                             0.002     0.002
</code></pre>
<pre class="output"><code>                                    NGenes  PropDown     PropUp Direction
WILLERT_WNT_SIGNALING                   22 0.2272727 0.54545455        Up
LABBE_WNT3A_TARGETS_UP                 104 0.2307692 0.49038462        Up
SANSOM_WNT_PATHWAY_REQUIRE_MYC          53 0.1886792 0.45283019        Up
PID_WNT_NONCANONICAL_PATHWAY            31 0.4516129 0.09677419      Down
REACTOME_SIGNALING_BY_WNT               62 0.1290323 0.50000000        Up
PID_WNT_CANONICAL_PATHWAY               19 0.4736842 0.21052632      Down
BIOCARTA_WNT_PATHWAY                    24 0.2500000 0.33333333        Up
LABBE_TARGETS_OF_TGFB1_AND_WNT3A_UP    100 0.3400000 0.38000000        Up
KEGG_WNT_SIGNALING_PATHWAY             131 0.3740458 0.25190840      Down
ST_WNT_BETA_CATENIN_PATHWAY             32 0.3125000 0.34375000        Up
PID_WNT_SIGNALING_PATHWAY               23 0.3043478 0.30434783        Up
LABBE_TARGETS_OF_TGFB1_AND_WNT3A_DN     90 0.2555556 0.40000000        Up
ST_WNT_CA2_CYCLIC_GMP_PATHWAY           14 0.2857143 0.35714286        Up
LABBE_WNT3A_TARGETS_DN                  65 0.2615385 0.27692308        Up
WNT_SIGNALING                           75 0.3600000 0.30666667      Down
                                    PValue        FDR PValue.Mixed
WILLERT_WNT_SIGNALING                0.001 0.00187500        0.001
LABBE_WNT3A_TARGETS_UP               0.001 0.00187500        0.001
SANSOM_WNT_PATHWAY_REQUIRE_MYC       0.001 0.00187500        0.001
PID_WNT_NONCANONICAL_PATHWAY         0.001 0.00187500        0.001
REACTOME_SIGNALING_BY_WNT            0.023 0.05625000        0.003
PID_WNT_CANONICAL_PATHWAY            0.023 0.05625000        0.001
BIOCARTA_WNT_PATHWAY                 0.040 0.08464286        0.001
LABBE_TARGETS_OF_TGFB1_AND_WNT3A_UP  0.051 0.08583333        0.001
KEGG_WNT_SIGNALING_PATHWAY           0.052 0.08583333        0.001
ST_WNT_BETA_CATENIN_PATHWAY          0.062 0.09225000        0.001
PID_WNT_SIGNALING_PATHWAY            0.103 0.13977273        0.001
LABBE_TARGETS_OF_TGFB1_AND_WNT3A_DN  0.163 0.20312500        0.001
ST_WNT_CA2_CYCLIC_GMP_PATHWAY        0.254 0.29250000        0.001
LABBE_WNT3A_TARGETS_DN               0.500 0.53517857        0.001
WNT_SIGNALING                        0.888 0.88800000        0.001
                                    FDR.Mixed
WILLERT_WNT_SIGNALING                   0.001
LABBE_WNT3A_TARGETS_UP                  0.001
SANSOM_WNT_PATHWAY_REQUIRE_MYC          0.001
PID_WNT_NONCANONICAL_PATHWAY            0.001
REACTOME_SIGNALING_BY_WNT               0.003
PID_WNT_CANONICAL_PATHWAY               0.001
BIOCARTA_WNT_PATHWAY                    0.001
LABBE_TARGETS_OF_TGFB1_AND_WNT3A_UP     0.001
KEGG_WNT_SIGNALING_PATHWAY              0.001
ST_WNT_BETA_CATENIN_PATHWAY             0.001
PID_WNT_SIGNALING_PATHWAY               0.001
LABBE_TARGETS_OF_TGFB1_AND_WNT3A_DN     0.001
ST_WNT_CA2_CYCLIC_GMP_PATHWAY           0.001
LABBE_WNT3A_TARGETS_DN                  0.001
WNT_SIGNALING                           0.001
</code></pre>
<p>Notes</p>
<ul>
<li>A common application of ROAST is to use a set of DE genes that was defined from an analysis of an independent data set. ROAST can then determine whether similar changes are observed in the contrast of interest for the current data set.</li>
<li>Even for GO-defined gene sets, goana and ROAST have different behaviours. In goana, the significance of differential expression for a GO term is determined relative to other DE genes that are not annotated with that term. In ROAST, only differential expression for the genes in the set are relevant to the significance of that set and its corresponding term. goana depends on a significance cutoff to choose DE genes, whereas ROAST does not require a cutoff and evaluates all genes in the set.</li>
<li>ROAST estimates p-values by simulation, so the results may change slightly between runs. More precise p-values can be obtained by increasing the number of rotations, albeit at the cost of increased computational time.</li>
<li>The smallest p-value that can be reported is 1/(2nrot + 1) where nrot is the number of rotations. This lower bound can be decreased by increasing nrot.</li>
</ul>
<h3 id="visualising-gene-set-tests-barcode-and-enrichment-plots">Visualising gene set tests: Barcode and enrichment plots</h3>
<p>A barcode plot can be produced with the <code>barcodeplot</code> function to visualize the results for any particular set. To display a barcodeplot, we need to decide what statistics to use, usually we choose either logFC or t-statistics for the comparison of interest. We also need to make sure we give the statistics in the correct order, such that subsetting the statistics vector will give the genes for the gene set we’re interested in. The best way to do this is to work with the fit object directly. The <code>coefficients</code> slot contains the logFCs and the <code>t</code> slot contains the t-statistics.</p>
<p>Let’s have a look at one of the top MYC related pathway in the ROAST test that you have already done, “MENSSEN_MYC_TARGETS”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Have a look at the logFCs and t-statistics in fit.cont</span>
<span class="kw">names</span>(fit.cont)</code></pre></div>
<pre class="output"><code> [1] &quot;coefficients&quot;     &quot;stdev.unscaled&quot;   &quot;sigma&quot;           
 [4] &quot;df.residual&quot;      &quot;cov.coefficients&quot; &quot;rank&quot;            
 [7] &quot;Amean&quot;            &quot;method&quot;           &quot;design&quot;          
[10] &quot;contrasts&quot;        &quot;df.prior&quot;         &quot;s2.prior&quot;        
[13] &quot;var.prior&quot;        &quot;proportion&quot;       &quot;s2.post&quot;         
[16] &quot;t&quot;                &quot;df.total&quot;         &quot;p.value&quot;         
[19] &quot;lods&quot;             &quot;F&quot;                &quot;F.p.value&quot;       
[22] &quot;genes&quot;           
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(fit.cont$coefficients)</code></pre></div>
<pre class="output"><code>        Contrasts
         B.PregVsLac L.PregVsLac
  497097 -1.62084708  -0.6964826
  20671   0.41298109   0.3652171
  27395  -0.03540567  -0.3849389
  18777  -0.08505647  -0.4731888
  21399   0.21912852   0.2839044
  58175   1.49852621   1.4736067
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(fit.cont$t)</code></pre></div>
<pre class="output"><code>        Contrasts
         B.PregVsLac L.PregVsLac
  497097  -4.0541969  -0.3076765
  20671    1.0030050   0.4055901
  27395   -0.1936036  -2.2395329
  18777   -0.7975228  -4.0746018
  21399    2.1866392   2.5284968
  58175    1.5788218   3.1387631
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="co"># barcode plot with logFCs</span>
<span class="kw">barcodeplot</span>(fit.cont$coeff[,<span class="dv">1</span>], <span class="dt">index=</span>c2.ind[[<span class="st">&quot;MENSSEN_MYC_TARGETS&quot;</span>]], <span class="dt">main=</span><span class="st">&quot;LogFC: MENSSEN_MYC_TARGETS&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-77-1.png" title="plot of chunk unnamed-chunk-77" alt="plot of chunk unnamed-chunk-77" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># barcode plot using t-statistics</span>
<span class="kw">barcodeplot</span>(fit.cont$t[,<span class="dv">1</span>], <span class="dt">index=</span>c2.ind[[<span class="st">&quot;MENSSEN_MYC_TARGETS&quot;</span>]], <span class="dt">main=</span><span class="st">&quot;T-statistic: MENSSEN_MYC_TARGETS&quot;</span>)</code></pre></div>
<p><img src="fig-06/unnamed-chunk-77-2.png" title="plot of chunk unnamed-chunk-77" alt="plot of chunk unnamed-chunk-77" style="display: block; margin: auto;" /></p>
<p>Here, genes are represented by bars and are ranked from left to right by decreasing log-fold change or t-statistic. This forms the barcode-like pattern. The line above the barcode shows the relative local enrichment of the vertical bars in each part of the plot. The barcodeplot shows that the genes in this gene set tend to be up-regulated between pregnant and lactating in basal cells.</p>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="challenge-11"><span class="glyphicon glyphicon-pencil"></span>Challenge</h2>
</div>
<div class="panel-body">
<ol style="list-style-type: decimal">
<li>Produce a barcodeplot for luminal pregnant vs lactating for this gene set. Does the pattern of enrichment look as strong?</li>
<li>Choose one of the Wnt signalling pathways and produce a barcode plot for both comparisons.</li>
<li>You can put two gene sets on one plot, for example a set that is up-regulated and one that is down-regulated, by adding a gene set to the <code>index2</code> argument. Produce a barcodeplot with two sets of your choosing.</li>
</ol>
</div>
</section>
<p><strong>Solution</strong> <img src="fig-06/unnamed-chunk-78-1.png" title="plot of chunk unnamed-chunk-78" alt="plot of chunk unnamed-chunk-78" style="display: block; margin: auto;" /><img src="fig-06/unnamed-chunk-78-2.png" title="plot of chunk unnamed-chunk-78" alt="plot of chunk unnamed-chunk-78" style="display: block; margin: auto;" /><img src="fig-06/unnamed-chunk-78-3.png" title="plot of chunk unnamed-chunk-78" alt="plot of chunk unnamed-chunk-78" style="display: block; margin: auto;" /></p>
<h1 id="record-package-and-version-info-with-sessioninfo">Record package and version info with sessionInfo()</h1>
<p>The session information describes the versions of R and of the packages that were used in the analysis. This is useful for record-keeping purposes, and ensures that an analysis can be reproduced even when the software is updated over time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre class="output"><code>R version 3.3.2 (2016-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X El Capitan 10.11.6

locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8

attached base packages:
[1] parallel  stats4    methods   stats     graphics  grDevices utils    
[8] datasets  base     

other attached packages:
 [1] RColorBrewer_1.1-2   org.Mm.eg.db_3.4.0   AnnotationDbi_1.36.0
 [4] IRanges_2.8.1        S4Vectors_0.12.0     Biobase_2.34.0      
 [7] BiocGenerics_0.20.0  gplots_3.0.1         Glimma_1.2.1        
[10] edgeR_3.16.3         limma_3.30.4         knitr_1.15          

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.4.0 genefilter_1.56.0         
 [3] gtools_3.5.0               locfit_1.5-9.1            
 [5] splines_3.3.2              lattice_0.20-34           
 [7] colorspace_1.3-1           htmltools_0.3.5           
 [9] chron_2.3-47               survival_2.40-1           
[11] XML_3.98-1.5               foreign_0.8-67            
[13] DBI_0.5-1                  BiocParallel_1.8.1        
[15] BiasedUrn_1.07             plyr_1.8.4                
[17] stringr_1.1.0              zlibbioc_1.20.0           
[19] munsell_0.4.3              gtable_0.2.0              
[21] caTools_1.17.1             DESeq2_1.14.0             
[23] evaluate_0.10              latticeExtra_0.6-28       
[25] geneplotter_1.52.0         GenomeInfoDb_1.10.1       
[27] highr_0.6                  htmlTable_1.7             
[29] Rcpp_0.12.8                KernSmooth_2.23-15        
[31] acepack_1.4.1              xtable_1.8-2              
[33] scales_0.4.1               gdata_2.17.0              
[35] Hmisc_4.0-0                annotate_1.52.0           
[37] XVector_0.14.0             gridExtra_2.2.1           
[39] ggplot2_2.2.0              digest_0.6.10             
[41] stringi_1.1.2              GenomicRanges_1.26.1      
[43] grid_3.3.2                 tools_3.3.2               
[45] bitops_1.0-6               magrittr_1.5              
[47] lazyeval_0.2.0             RCurl_1.95-4.8            
[49] tibble_1.2                 RSQLite_1.0.0             
[51] Formula_1.2-1              cluster_2.0.5             
[53] GO.db_3.4.0                Matrix_1.2-7.1            
[55] data.table_1.9.6           assertthat_0.1            
[57] rpart_4.1-10               nnet_7.3-12               
</code></pre>
<h1 id="references" class="unnumbered">References</h1>
<div id="refs" class="references">
<div id="ref-Fu2015">
<p>Fu, Nai Yang, Anne C Rios, Bhupinder Pal, Rina Soetanto, Aaron T L Lun, Kevin Liu, Tamara Beck, et al. 2015. “EGF-mediated induction of Mcl-1 at the switch to lactation is essential for alveolar cell survival.” <em>Nature Cell Biology</em> 17 (4): 365–75. doi:<a href="https://doi.org/10.1038/ncb3117">10.1038/ncb3117</a>.</p>
</div>
<div id="ref-Law2014">
<p>Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom: precision weights unlock linear model analysis tools for RNA-seq read counts.” <em>Genome Biology</em> 15 (2): R29. doi:<a href="https://doi.org/10.1186/gb-2014-15-2-r29">10.1186/gb-2014-15-2-r29</a>.</p>
</div>
<div id="ref-Liao2014">
<p>Liao, Yang, Gordon K Smyth, and Wei Shi. 2014. “featureCounts: an efficient general purpose program for assigning sequence reads to genomic features.” <em>Bioinformatics (Oxford, England)</em> 30 (7): 923–30. doi:<a href="https://doi.org/10.1093/bioinformatics/btt656">10.1093/bioinformatics/btt656</a>.</p>
</div>
<div id="ref-Lun2016">
<p>Lun, Aaron T L, Yunshun Chen, and Gordon K Smyth. 2016. “It’s DE-licious: A Recipe for Differential Expression Analyses of RNA-seq Experiments Using Quasi-Likelihood Methods in edgeR.” <em>Methods in Molecular Biology (Clifton, N.J.)</em> 1418 (January): 391–416. doi:<a href="https://doi.org/10.1007/978-1-4939-3578-9\_19">10.1007/978-1-4939-3578-9\_19</a>.</p>
</div>
<div id="ref-McCarthy2009">
<p>McCarthy, Davis J, and Gordon K Smyth. 2009. “Testing significance relative to a fold-change threshold is a TREAT.” <em>Bioinformatics (Oxford, England)</em> 25 (6): 765–71. doi:<a href="https://doi.org/10.1093/bioinformatics/btp053">10.1093/bioinformatics/btp053</a>.</p>
</div>
<div id="ref-Ritchie2015">
<p>Ritchie, M. E., B. Phipson, D. Wu, Y. Hu, C. W. Law, W. Shi, and G. K. Smyth. 2015. “limma powers differential expression analyses for RNA-sequencing and microarray studies.” <em>Nucleic Acids Research</em>, January, gkv007. doi:<a href="https://doi.org/10.1093/nar/gkv007">10.1093/nar/gkv007</a>.</p>
</div>
<div id="ref-robinson2010edgeR">
<p>Robinson, M D, D J McCarthy, and G K Smyth. 2010. “edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.” <em>Bioinformatics</em> 26 (1). Oxford Univ Press: 139–40.</p>
</div>
<div id="ref-robinson2010tmm">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A scaling normalization method for differential expression analysis of RNA-seq data.” <em>Genome Biology</em> 11 (3): R25. doi:<a href="https://doi.org/10.1186/gb-2010-11-3-r25">10.1186/gb-2010-11-3-r25</a>.</p>
</div>
<div id="ref-wu2012camera">
<p>Wu, D, and G K Smyth. 2012. “Camera: a competitive gene set test accounting for inter-gene correlation.” <em>Nucleic Acids Research</em> 40 (17). Oxford University Press: e133—–e133.</p>
</div>
<div id="ref-wu2010roast">
<p>Wu, D, E Lim, F Vaillant, M L Asselin-Labat, J E Visvader, and G K Smyth. 2010. “ROAST: rotation gene set tests for complex microarray experiments.” <em>Bioinformatics</em> 26 (17). Oxford Univ Press: 2176–82.</p>
</div>
</div>
        </div>
      </div>
      </article>
      
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
