---
title: "RNA-seq analysis in R"
subtitle: "Differential Expression of RNA-seq data"
author: "Stephane Ballereau, Mark Dunning, Abbi Edwards, Oscar Rueda, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
  html_document:
    toc: yes
minutes: 300
layout: page
bibliography: ref.bib
editor_options: 
  chunk_output_type: inline
---

# Recap of pre-processing

The previous section walked-through the pre-processing and transformation of the
count data. Here, for completeness, we list the minimal steps required to 
process the data prior to differential expression analysis.

```{r recap, eval = FALSE}
# Read the sample information into a data frame
sampleinfo <- read_tsv("data/SampleInfo.Corrected.txt")

# Read the data into R
seqdata <- read_tsv("data/GSE60450_Lactation.featureCounts", comment = "#")

# Transform the data to matrix of counts
countdata <- as.data.frame(seqdata) %>% 
    column_to_rownames("Geneid") %>% # turn the geneid column into rownames
    rename_all(str_remove, ".bam") %>% # remove the ".bam" from the column names
    select(sampleinfo$Sample) %>% # keep sample columns using sampleinfo
    as.matrix()

# filter the data to remove genes with few counts
keep <- rowSums(countdata) > 5
countdata <- countdata[keep,]
```

## Load the data

Alternatively we can load the `objects with the RData file we created in the 
pre-processing tutorial.

First load the packages we need.

```{r setup, message = FALSE}
library(tidyverse)
library(DESeq2)
```

```{r loadData}
# load the RData object we created in the previous session
load("Robjects/preprocessing.RData")
ls()
dim(countdata)
sampleinfo
```

# The model formula and design matrices

Now that we are happy that the quality of the data looks good, we can 
proceed to testing for differentially expressed genes. There are a number of 
packages to analyse RNA-Seq data. Most people use 
[DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 
or [edgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html). They 
are both equally applicable. There is an informative and honest blog post
[here](https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger/) by Mike Love,
one of the authors of DESeq2, about deciding which to use.

We will use **DESeq2** for the rest of this practical.

## Create a DESeqDataSet object with the raw data

### Creating the design model formula

First we need to create a design model formula for our analysis. `DESeq2` will 
use this to generate the model matrix, as we have seen in the linear models 
lecture. 

We have two variables: "status"" and "cell type". We will fit two models under 
two assumptions: no interaction and interaction of these two factors. 

Let's start with the model with only main effects, that is no interaction. 
The main assumption here is that the effect of the status is the same in both 
type of cells.

```{r modelForumla}
# Use the standard R 'formula' syntax for an additive model
design <- as.formula(~ CellType + Status)
```

What does this look like as a model matrix?
```{r modelMatrix}
modelMatrix <- model.matrix(design, data = sampleinfo)
modelMatrix
```

It would be nice if `virgin` were the base line/intercept. To get R to 
use `virgin` as the intercept we need to use a `factor`. Let's set factor levels
on Status to use virgin as the intercept.

```{r setFactors}
sampleinfo$Status <- factor(sampleinfo$Status, 
                            levels = c("virgin", "pregnant", "lactate"))
modelMatrix <- model.matrix(design, data = sampleinfo)
modelMatrix
```

# Build a DESeq2DataSet

We don't actually need to pass `DESeq2` the model matrix, instead we pass it the 
design formula and the `sampleinfo` it will build the matrix itself.

```{r makeDDSObj}
# create the DESeqDataSet object
ddsObj.raw <- DESeqDataSetFromMatrix(countData = countdata,
                                     colData = sampleinfo,
                                     design = design)
```

# Data exploration

Let's plot a PCA from `vst` transformed data. 
Can you anticipate if the interaction term will be important?

```{r pcaPlot, fig.width=5, fig.height=5}
vstcounts <- vst(ddsObj.raw, blind=TRUE)
plotPCA(vstcounts, intgroup=c("Status", "CellType"))
```

# Differential expression analysis with DESeq2

## The `DESeq2` work flow

The main `DESeq2` work flow is carried out in 3 steps:

First, Calculate the "median ratio" normalisation size factors...

```{r commonSizeFactors}
ddsObj <- estimateSizeFactors(ddsObj.raw)
```

... then estimate dispersion ...

```{r genewiseDispersion}
ddsObj <- estimateDispersions(ddsObj)
```

... finally, apply Negative Binomial GLM fitting and calculate Wald statistics
```{r applyGLM}
ddsObj <- nbinomWaldTest(ddsObj)
```

## The `DESeq` command

In practice the 3 steps above can be performed in a single step using the 
`DESeq` wrapper function. Performing the three steps separately is useful if you
wish to alter the default parameters of one or more steps, otherwise the `DESeq`
function is fine.

```{r theShortVersion}
# Run DESeq
ddsObj <- DESeq(ddsObj.raw)
```

## Generate a results table

We can generate a table of differential expression results from the DDS object
using the `results` function of DESeq2.

```{r resultsTable}
res <- results(ddsObj, alpha=0.05)
head(res)
```

### Independent filtering

You will notice that some of the adjusted p-values (`padj`) are NA. Remember 
in Session 2 we said that there is no need to pre-filter the genes as DESeq2
will do this through a process it calls 'independent filtering'. The genes 
with `NA` are the ones `DESeq2` has filtered out.

From `DESeq2` manual:
"The results function of the `DESeq2` package performs independent filtering by
default using the mean of normalized counts as a filter statistic. A threshold 
on the filter statistic is found which optimizes the number of adjusted p values
lower than a [specified] significance level".

The default significance level for independent filtering is `0.1`, however, you
should set this to the FDR cut off you are planning to use. We will use `0.05` -
this was the purpose of the `alpha` argument in the previous command.

### The default contrast of `results`

The `results` function has returned the results for the contrast "lactate vs 
virgin". Let's have a look at the model matrix to understand why `DESeq2` has 
given us this particular contrast.

```{r viewModelMatrix}
modelMatrix
```

By default, `results` has returned the contrast encoded by the final column in
the model matrix. `DESeq2` has the command `resultsNames` that allows us to
view the contrasts that are available directly from the model matrix.

```{r resultsNames}
resultsNames(ddsObj)
```

Let's just rename `res` so that we know which contrast results it contains.

```{r}
resLvV <- res
rm(res)
```

If we want a different contrast we can just pass the `results` function the name
of the design matrix column that encodes it.
Let's retrieve the results for pregant versus virgin

```{r resultPvV}
resPvV <- results(ddsObj, 
                  name="Status_pregnant_vs_virgin", 
                  alpha = 0.05)
resPvV
```

Let's get the top 100 genes by adjusted p-value

```{r topGenesPvV, message = F, results="hide"}
topGenesPvV <- as.data.frame(resPvV) %>%
    rownames_to_column("GeneID") %>% 
    arrange(padj) %>% 
    head(100)
topGenesPvV
```

```{r topGenesPvVprint, echo=FALSE}
head(topGenesPvV)
```


> #### Challenge 1 {.challenge}
> Obtain results for luminal vs basal and find the top 200 genes.
> Call the new results object `resBvL`.

```{r solutionChallenge1}
```

## Contrasts

Suppose we want to find differentially expressed genes between **pregnant** and 
**lactate**. We don't have a parameter that explicitly will allow us to test 
that hypothesis. We need to provide a contrast.

```{r makeContrast}
resultsNames(ddsObj)

resPvL <- results(ddsObj,
                  contrast=c("Status", "pregnant", "lactate"), 
                  alpha = 0.05)
resPvL
```

# Comparing two design models

Suppose we thought that maybe `status` were irrelevant and really the only 
differences might be between cell types. We could fit a simpler model, this 
would give us more degrees of freedom and therefore more power, but how
would we know if it was a better model of not? We can compare the two models
using the "log ratio test" (LRT).

```{r compareModels}
designC <- as.formula(~ CellType )

# Compare the designs
ddsObjC <- DESeq(ddsObj, test="LRT", reduced=designC)
resCvCS <- results(ddsObjC)
resCvCS
```

The null hypothesis is that there is no significant difference between the two
models, i.e. the simpler model is sufficient to explain the variation in gene
expression between the samples. If the the adjusted p-value for a gene passes
a significance threshold (e.g. padj < 0.05) then we should consider using the 
more complex model for this gene. In practice we would usually choose one model
or the other and apply it to all genes.

> ### Challenge 2 {.challenge}
> When we looked at the PCA it did seem that an interaction model might be
> warranted. Let's test that.  
>  
> 1. Fit a model with interaction.  
> 2. Use the LRT to compare the two models.  
> 3. Is the number of replicates good enough to include the interaction?    
> 4. Is the interaction needed in the model?  

```{r solutionChallenge2}
```

## Finally save the results in a new RData object

```{r saveObjects, eval=FALSE}
save(resLvV, ddsObj, sampleinfo, file="results/DE.RData")
```

--------------------

